// @microsoft/teamsfx-core/build/component/driver/teamsApp/interfaces/appdefinitions/appCategory.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppCategory = void 0;
var AppCategory;
(function (AppCategory) {
    AppCategory[AppCategory["Analytics"] = 0] = "Analytics";
    AppCategory[AppCategory["Communication"] = 1] = "Communication";
    AppCategory[AppCategory["ContentManagement"] = 2] = "ContentManagement";
    AppCategory[AppCategory["CustomerAndContactManagement"] = 3] = "CustomerAndContactManagement";
    AppCategory[AppCategory["CustomerSupport"] = 4] = "CustomerSupport";
    AppCategory[AppCategory["DataVisualizationAndBI"] = 5] = "DataVisualizationAndBI";
    AppCategory[AppCategory["DeveloperTools"] = 6] = "DeveloperTools";
    AppCategory[AppCategory["Education"] = 7] = "Education";
    AppCategory[AppCategory["FilesAndDocuments"] = 8] = "FilesAndDocuments";
    AppCategory[AppCategory["FinancialManagement"] = 9] = "FinancialManagement";
    AppCategory[AppCategory["HumanResourcesAndRecruiting"] = 10] = "HumanResourcesAndRecruiting";
    AppCategory[AppCategory["ImageAndVideoGalleries"] = 11] = "ImageAndVideoGalleries";
    AppCategory[AppCategory["ITAdmin"] = 12] = "ITAdmin";
    AppCategory[AppCategory["Jobs"] = 13] = "Jobs";
    AppCategory[AppCategory["LabApp"] = 14] = "LabApp";
    AppCategory[AppCategory["Lifestyle"] = 15] = "Lifestyle";
    AppCategory[AppCategory["LookAndFeel"] = 16] = "LookAndFeel";
    AppCategory[AppCategory["MapsAndFeeds"] = 17] = "MapsAndFeeds";
    AppCategory[AppCategory["MeetingsAndScheduling"] = 18] = "MeetingsAndScheduling";
    AppCategory[AppCategory["Microsoft"] = 19] = "Microsoft";
    AppCategory[AppCategory["NewsAndWeather"] = 20] = "NewsAndWeather";
    AppCategory[AppCategory["Other"] = 21] = "Other";
    AppCategory[AppCategory["PrintingAndFaxing"] = 22] = "PrintingAndFaxing";
    AppCategory[AppCategory["Productivity"] = 23] = "Productivity";
    AppCategory[AppCategory["ProjectManagement"] = 24] = "ProjectManagement";
    AppCategory[AppCategory["PublicFacingWebsites"] = 25] = "PublicFacingWebsites";
    AppCategory[AppCategory["Reference"] = 26] = "Reference";
    AppCategory[AppCategory["SalesAndMarketing"] = 27] = "SalesAndMarketing";
    AppCategory[AppCategory["SearchAndDiscoverability"] = 28] = "SearchAndDiscoverability";
    AppCategory[AppCategory["SecurityAndCompliance"] = 29] = "SecurityAndCompliance";
    AppCategory[AppCategory["Social"] = 30] = "Social";
    AppCategory[AppCategory["TrainingAndTutorial"] = 31] = "TrainingAndTutorial";
    AppCategory[AppCategory["TravelAndNavigation"] = 32] = "TravelAndNavigation";
    AppCategory[AppCategory["Utilities"] = 33] = "Utilities";
    AppCategory[AppCategory["WorkflowAndBusinessManagement"] = 34] = "WorkflowAndBusinessManagement";
})(AppCategory = exports.AppCategory || (exports.AppCategory = {}));
