// @microsoft/teamsfx-core/build/component/driver/teamsApp/interfaces/appdefinitions/appItemIndustry.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppItemIndustry = void 0;
var AppItemIndustry;
(function (AppItemIndustry) {
    AppItemIndustry[AppItemIndustry["Agriculture"] = 0] = "Agriculture";
    AppItemIndustry[AppItemIndustry["Distribution"] = 1] = "Distribution";
    AppItemIndustry[AppItemIndustry["Education"] = 2] = "Education";
    AppItemIndustry[AppItemIndustry["Finance"] = 3] = "Finance";
    AppItemIndustry[AppItemIndustry["Government"] = 4] = "Government";
    AppItemIndustry[AppItemIndustry["HealthCareAndLifeSciences"] = 5] = "HealthCareAndLifeSciences";
    AppItemIndustry[AppItemIndustry["Manufacturing"] = 6] = "Manufacturing";
    AppItemIndustry[AppItemIndustry["Other"] = 7] = "Other";
    AppItemIndustry[AppItemIndustry["ProfessionalServices"] = 8] = "ProfessionalServices";
    AppItemIndustry[AppItemIndustry["Retail"] = 9] = "Retail";
})(AppItemIndustry = exports.AppItemIndustry || (exports.AppItemIndustry = {}));
