// @azure/arm-sql/dist-esm/src/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link GeoBackupPolicyName} that the service accepts. */
export var KnownGeoBackupPolicyName;
(function (KnownGeoBackupPolicyName) {
    /** Default */
    KnownGeoBackupPolicyName["Default"] = "Default";
})(KnownGeoBackupPolicyName || (KnownGeoBackupPolicyName = {}));
/** Known values of {@link UnitType} that the service accepts. */
export var KnownUnitType;
(function (KnownUnitType) {
    /** Count */
    KnownUnitType["Count"] = "count";
    /** Bytes */
    KnownUnitType["Bytes"] = "bytes";
    /** Seconds */
    KnownUnitType["Seconds"] = "seconds";
    /** Percent */
    KnownUnitType["Percent"] = "percent";
    /** CountPerSecond */
    KnownUnitType["CountPerSecond"] = "countPerSecond";
    /** BytesPerSecond */
    KnownUnitType["BytesPerSecond"] = "bytesPerSecond";
})(KnownUnitType || (KnownUnitType = {}));
/** Known values of {@link PrimaryAggregationType} that the service accepts. */
export var KnownPrimaryAggregationType;
(function (KnownPrimaryAggregationType) {
    /** None */
    KnownPrimaryAggregationType["None"] = "None";
    /** Average */
    KnownPrimaryAggregationType["Average"] = "Average";
    /** Count */
    KnownPrimaryAggregationType["Count"] = "Count";
    /** Minimum */
    KnownPrimaryAggregationType["Minimum"] = "Minimum";
    /** Maximum */
    KnownPrimaryAggregationType["Maximum"] = "Maximum";
    /** Total */
    KnownPrimaryAggregationType["Total"] = "Total";
})(KnownPrimaryAggregationType || (KnownPrimaryAggregationType = {}));
/** Known values of {@link UnitDefinitionType} that the service accepts. */
export var KnownUnitDefinitionType;
(function (KnownUnitDefinitionType) {
    /** Count */
    KnownUnitDefinitionType["Count"] = "Count";
    /** Bytes */
    KnownUnitDefinitionType["Bytes"] = "Bytes";
    /** Seconds */
    KnownUnitDefinitionType["Seconds"] = "Seconds";
    /** Percent */
    KnownUnitDefinitionType["Percent"] = "Percent";
    /** CountPerSecond */
    KnownUnitDefinitionType["CountPerSecond"] = "CountPerSecond";
    /** BytesPerSecond */
    KnownUnitDefinitionType["BytesPerSecond"] = "BytesPerSecond";
})(KnownUnitDefinitionType || (KnownUnitDefinitionType = {}));
/** Known values of {@link RecommendedActionCurrentState} that the service accepts. */
export var KnownRecommendedActionCurrentState;
(function (KnownRecommendedActionCurrentState) {
    /** Active */
    KnownRecommendedActionCurrentState["Active"] = "Active";
    /** Pending */
    KnownRecommendedActionCurrentState["Pending"] = "Pending";
    /** Executing */
    KnownRecommendedActionCurrentState["Executing"] = "Executing";
    /** Verifying */
    KnownRecommendedActionCurrentState["Verifying"] = "Verifying";
    /** PendingRevert */
    KnownRecommendedActionCurrentState["PendingRevert"] = "PendingRevert";
    /** RevertCancelled */
    KnownRecommendedActionCurrentState["RevertCancelled"] = "RevertCancelled";
    /** Reverting */
    KnownRecommendedActionCurrentState["Reverting"] = "Reverting";
    /** Reverted */
    KnownRecommendedActionCurrentState["Reverted"] = "Reverted";
    /** Ignored */
    KnownRecommendedActionCurrentState["Ignored"] = "Ignored";
    /** Expired */
    KnownRecommendedActionCurrentState["Expired"] = "Expired";
    /** Monitoring */
    KnownRecommendedActionCurrentState["Monitoring"] = "Monitoring";
    /** Resolved */
    KnownRecommendedActionCurrentState["Resolved"] = "Resolved";
    /** Success */
    KnownRecommendedActionCurrentState["Success"] = "Success";
    /** Error */
    KnownRecommendedActionCurrentState["Error"] = "Error";
})(KnownRecommendedActionCurrentState || (KnownRecommendedActionCurrentState = {}));
/** Known values of {@link ColumnDataType} that the service accepts. */
export var KnownColumnDataType;
(function (KnownColumnDataType) {
    /** Image */
    KnownColumnDataType["Image"] = "image";
    /** Text */
    KnownColumnDataType["Text"] = "text";
    /** Uniqueidentifier */
    KnownColumnDataType["Uniqueidentifier"] = "uniqueidentifier";
    /** Date */
    KnownColumnDataType["Date"] = "date";
    /** Time */
    KnownColumnDataType["Time"] = "time";
    /** Datetime2 */
    KnownColumnDataType["Datetime2"] = "datetime2";
    /** Datetimeoffset */
    KnownColumnDataType["Datetimeoffset"] = "datetimeoffset";
    /** Tinyint */
    KnownColumnDataType["Tinyint"] = "tinyint";
    /** Smallint */
    KnownColumnDataType["Smallint"] = "smallint";
    /** Int */
    KnownColumnDataType["Int"] = "int";
    /** Smalldatetime */
    KnownColumnDataType["Smalldatetime"] = "smalldatetime";
    /** Real */
    KnownColumnDataType["Real"] = "real";
    /** Money */
    KnownColumnDataType["Money"] = "money";
    /** Datetime */
    KnownColumnDataType["Datetime"] = "datetime";
    /** Float */
    KnownColumnDataType["Float"] = "float";
    /** SqlVariant */
    KnownColumnDataType["SqlVariant"] = "sql_variant";
    /** Ntext */
    KnownColumnDataType["Ntext"] = "ntext";
    /** Bit */
    KnownColumnDataType["Bit"] = "bit";
    /** Decimal */
    KnownColumnDataType["Decimal"] = "decimal";
    /** Numeric */
    KnownColumnDataType["Numeric"] = "numeric";
    /** Smallmoney */
    KnownColumnDataType["Smallmoney"] = "smallmoney";
    /** Bigint */
    KnownColumnDataType["Bigint"] = "bigint";
    /** Hierarchyid */
    KnownColumnDataType["Hierarchyid"] = "hierarchyid";
    /** Geometry */
    KnownColumnDataType["Geometry"] = "geometry";
    /** Geography */
    KnownColumnDataType["Geography"] = "geography";
    /** Varbinary */
    KnownColumnDataType["Varbinary"] = "varbinary";
    /** Varchar */
    KnownColumnDataType["Varchar"] = "varchar";
    /** Binary */
    KnownColumnDataType["Binary"] = "binary";
    /** Char */
    KnownColumnDataType["Char"] = "char";
    /** Timestamp */
    KnownColumnDataType["Timestamp"] = "timestamp";
    /** Nvarchar */
    KnownColumnDataType["Nvarchar"] = "nvarchar";
    /** Nchar */
    KnownColumnDataType["Nchar"] = "nchar";
    /** Xml */
    KnownColumnDataType["Xml"] = "xml";
    /** Sysname */
    KnownColumnDataType["Sysname"] = "sysname";
})(KnownColumnDataType || (KnownColumnDataType = {}));
/** Known values of {@link TableTemporalType} that the service accepts. */
export var KnownTableTemporalType;
(function (KnownTableTemporalType) {
    /** NonTemporalTable */
    KnownTableTemporalType["NonTemporalTable"] = "NonTemporalTable";
    /** HistoryTable */
    KnownTableTemporalType["HistoryTable"] = "HistoryTable";
    /** SystemVersionedTemporalTable */
    KnownTableTemporalType["SystemVersionedTemporalTable"] = "SystemVersionedTemporalTable";
})(KnownTableTemporalType || (KnownTableTemporalType = {}));
/** Known values of {@link SecurityAlertPolicyName} that the service accepts. */
export var KnownSecurityAlertPolicyName;
(function (KnownSecurityAlertPolicyName) {
    /** Default */
    KnownSecurityAlertPolicyName["Default"] = "Default";
})(KnownSecurityAlertPolicyName || (KnownSecurityAlertPolicyName = {}));
/** Known values of {@link CreatedByType} that the service accepts. */
export var KnownCreatedByType;
(function (KnownCreatedByType) {
    /** User */
    KnownCreatedByType["User"] = "User";
    /** Application */
    KnownCreatedByType["Application"] = "Application";
    /** ManagedIdentity */
    KnownCreatedByType["ManagedIdentity"] = "ManagedIdentity";
    /** Key */
    KnownCreatedByType["Key"] = "Key";
})(KnownCreatedByType || (KnownCreatedByType = {}));
/** Known values of {@link VulnerabilityAssessmentName} that the service accepts. */
export var KnownVulnerabilityAssessmentName;
(function (KnownVulnerabilityAssessmentName) {
    /** Default */
    KnownVulnerabilityAssessmentName["Default"] = "default";
})(KnownVulnerabilityAssessmentName || (KnownVulnerabilityAssessmentName = {}));
/** Known values of {@link VulnerabilityAssessmentScanTriggerType} that the service accepts. */
export var KnownVulnerabilityAssessmentScanTriggerType;
(function (KnownVulnerabilityAssessmentScanTriggerType) {
    /** OnDemand */
    KnownVulnerabilityAssessmentScanTriggerType["OnDemand"] = "OnDemand";
    /** Recurring */
    KnownVulnerabilityAssessmentScanTriggerType["Recurring"] = "Recurring";
})(KnownVulnerabilityAssessmentScanTriggerType || (KnownVulnerabilityAssessmentScanTriggerType = {}));
/** Known values of {@link VulnerabilityAssessmentScanState} that the service accepts. */
export var KnownVulnerabilityAssessmentScanState;
(function (KnownVulnerabilityAssessmentScanState) {
    /** Passed */
    KnownVulnerabilityAssessmentScanState["Passed"] = "Passed";
    /** Failed */
    KnownVulnerabilityAssessmentScanState["Failed"] = "Failed";
    /** FailedToRun */
    KnownVulnerabilityAssessmentScanState["FailedToRun"] = "FailedToRun";
    /** InProgress */
    KnownVulnerabilityAssessmentScanState["InProgress"] = "InProgress";
})(KnownVulnerabilityAssessmentScanState || (KnownVulnerabilityAssessmentScanState = {}));
/** Known values of {@link DataWarehouseUserActivityName} that the service accepts. */
export var KnownDataWarehouseUserActivityName;
(function (KnownDataWarehouseUserActivityName) {
    /** Current */
    KnownDataWarehouseUserActivityName["Current"] = "current";
})(KnownDataWarehouseUserActivityName || (KnownDataWarehouseUserActivityName = {}));
/** Known values of {@link ElasticPoolState} that the service accepts. */
export var KnownElasticPoolState;
(function (KnownElasticPoolState) {
    /** Creating */
    KnownElasticPoolState["Creating"] = "Creating";
    /** Ready */
    KnownElasticPoolState["Ready"] = "Ready";
    /** Disabled */
    KnownElasticPoolState["Disabled"] = "Disabled";
})(KnownElasticPoolState || (KnownElasticPoolState = {}));
/** Known values of {@link ElasticPoolLicenseType} that the service accepts. */
export var KnownElasticPoolLicenseType;
(function (KnownElasticPoolLicenseType) {
    /** LicenseIncluded */
    KnownElasticPoolLicenseType["LicenseIncluded"] = "LicenseIncluded";
    /** BasePrice */
    KnownElasticPoolLicenseType["BasePrice"] = "BasePrice";
})(KnownElasticPoolLicenseType || (KnownElasticPoolLicenseType = {}));
/** Known values of {@link ServerKeyType} that the service accepts. */
export var KnownServerKeyType;
(function (KnownServerKeyType) {
    /** ServiceManaged */
    KnownServerKeyType["ServiceManaged"] = "ServiceManaged";
    /** AzureKeyVault */
    KnownServerKeyType["AzureKeyVault"] = "AzureKeyVault";
})(KnownServerKeyType || (KnownServerKeyType = {}));
/** Known values of {@link EncryptionProtectorName} that the service accepts. */
export var KnownEncryptionProtectorName;
(function (KnownEncryptionProtectorName) {
    /** Current */
    KnownEncryptionProtectorName["Current"] = "current";
})(KnownEncryptionProtectorName || (KnownEncryptionProtectorName = {}));
/** Known values of {@link ReadWriteEndpointFailoverPolicy} that the service accepts. */
export var KnownReadWriteEndpointFailoverPolicy;
(function (KnownReadWriteEndpointFailoverPolicy) {
    /** Manual */
    KnownReadWriteEndpointFailoverPolicy["Manual"] = "Manual";
    /** Automatic */
    KnownReadWriteEndpointFailoverPolicy["Automatic"] = "Automatic";
})(KnownReadWriteEndpointFailoverPolicy || (KnownReadWriteEndpointFailoverPolicy = {}));
/** Known values of {@link ReadOnlyEndpointFailoverPolicy} that the service accepts. */
export var KnownReadOnlyEndpointFailoverPolicy;
(function (KnownReadOnlyEndpointFailoverPolicy) {
    /** Disabled */
    KnownReadOnlyEndpointFailoverPolicy["Disabled"] = "Disabled";
    /** Enabled */
    KnownReadOnlyEndpointFailoverPolicy["Enabled"] = "Enabled";
})(KnownReadOnlyEndpointFailoverPolicy || (KnownReadOnlyEndpointFailoverPolicy = {}));
/** Known values of {@link FailoverGroupReplicationRole} that the service accepts. */
export var KnownFailoverGroupReplicationRole;
(function (KnownFailoverGroupReplicationRole) {
    /** Primary */
    KnownFailoverGroupReplicationRole["Primary"] = "Primary";
    /** Secondary */
    KnownFailoverGroupReplicationRole["Secondary"] = "Secondary";
})(KnownFailoverGroupReplicationRole || (KnownFailoverGroupReplicationRole = {}));
/** Known values of {@link InstanceFailoverGroupReplicationRole} that the service accepts. */
export var KnownInstanceFailoverGroupReplicationRole;
(function (KnownInstanceFailoverGroupReplicationRole) {
    /** Primary */
    KnownInstanceFailoverGroupReplicationRole["Primary"] = "Primary";
    /** Secondary */
    KnownInstanceFailoverGroupReplicationRole["Secondary"] = "Secondary";
})(KnownInstanceFailoverGroupReplicationRole || (KnownInstanceFailoverGroupReplicationRole = {}));
/** Known values of {@link InstancePoolLicenseType} that the service accepts. */
export var KnownInstancePoolLicenseType;
(function (KnownInstancePoolLicenseType) {
    /** LicenseIncluded */
    KnownInstancePoolLicenseType["LicenseIncluded"] = "LicenseIncluded";
    /** BasePrice */
    KnownInstancePoolLicenseType["BasePrice"] = "BasePrice";
})(KnownInstancePoolLicenseType || (KnownInstancePoolLicenseType = {}));
/** Known values of {@link JobAgentState} that the service accepts. */
export var KnownJobAgentState;
(function (KnownJobAgentState) {
    /** Creating */
    KnownJobAgentState["Creating"] = "Creating";
    /** Ready */
    KnownJobAgentState["Ready"] = "Ready";
    /** Updating */
    KnownJobAgentState["Updating"] = "Updating";
    /** Deleting */
    KnownJobAgentState["Deleting"] = "Deleting";
    /** Disabled */
    KnownJobAgentState["Disabled"] = "Disabled";
})(KnownJobAgentState || (KnownJobAgentState = {}));
/** Known values of {@link JobExecutionLifecycle} that the service accepts. */
export var KnownJobExecutionLifecycle;
(function (KnownJobExecutionLifecycle) {
    /** Created */
    KnownJobExecutionLifecycle["Created"] = "Created";
    /** InProgress */
    KnownJobExecutionLifecycle["InProgress"] = "InProgress";
    /** WaitingForChildJobExecutions */
    KnownJobExecutionLifecycle["WaitingForChildJobExecutions"] = "WaitingForChildJobExecutions";
    /** WaitingForRetry */
    KnownJobExecutionLifecycle["WaitingForRetry"] = "WaitingForRetry";
    /** Succeeded */
    KnownJobExecutionLifecycle["Succeeded"] = "Succeeded";
    /** SucceededWithSkipped */
    KnownJobExecutionLifecycle["SucceededWithSkipped"] = "SucceededWithSkipped";
    /** Failed */
    KnownJobExecutionLifecycle["Failed"] = "Failed";
    /** TimedOut */
    KnownJobExecutionLifecycle["TimedOut"] = "TimedOut";
    /** Canceled */
    KnownJobExecutionLifecycle["Canceled"] = "Canceled";
    /** Skipped */
    KnownJobExecutionLifecycle["Skipped"] = "Skipped";
})(KnownJobExecutionLifecycle || (KnownJobExecutionLifecycle = {}));
/** Known values of {@link ProvisioningState} that the service accepts. */
export var KnownProvisioningState;
(function (KnownProvisioningState) {
    /** Created */
    KnownProvisioningState["Created"] = "Created";
    /** InProgress */
    KnownProvisioningState["InProgress"] = "InProgress";
    /** Succeeded */
    KnownProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownProvisioningState["Failed"] = "Failed";
    /** Canceled */
    KnownProvisioningState["Canceled"] = "Canceled";
})(KnownProvisioningState || (KnownProvisioningState = {}));
/** Known values of {@link JobTargetType} that the service accepts. */
export var KnownJobTargetType;
(function (KnownJobTargetType) {
    /** TargetGroup */
    KnownJobTargetType["TargetGroup"] = "TargetGroup";
    /** SqlDatabase */
    KnownJobTargetType["SqlDatabase"] = "SqlDatabase";
    /** SqlElasticPool */
    KnownJobTargetType["SqlElasticPool"] = "SqlElasticPool";
    /** SqlShardMap */
    KnownJobTargetType["SqlShardMap"] = "SqlShardMap";
    /** SqlServer */
    KnownJobTargetType["SqlServer"] = "SqlServer";
})(KnownJobTargetType || (KnownJobTargetType = {}));
/** Known values of {@link JobStepActionType} that the service accepts. */
export var KnownJobStepActionType;
(function (KnownJobStepActionType) {
    /** TSql */
    KnownJobStepActionType["TSql"] = "TSql";
})(KnownJobStepActionType || (KnownJobStepActionType = {}));
/** Known values of {@link JobStepActionSource} that the service accepts. */
export var KnownJobStepActionSource;
(function (KnownJobStepActionSource) {
    /** Inline */
    KnownJobStepActionSource["Inline"] = "Inline";
})(KnownJobStepActionSource || (KnownJobStepActionSource = {}));
/** Known values of {@link JobStepOutputType} that the service accepts. */
export var KnownJobStepOutputType;
(function (KnownJobStepOutputType) {
    /** SqlDatabase */
    KnownJobStepOutputType["SqlDatabase"] = "SqlDatabase";
})(KnownJobStepOutputType || (KnownJobStepOutputType = {}));
/** Known values of {@link CapabilityGroup} that the service accepts. */
export var KnownCapabilityGroup;
(function (KnownCapabilityGroup) {
    /** SupportedEditions */
    KnownCapabilityGroup["SupportedEditions"] = "supportedEditions";
    /** SupportedElasticPoolEditions */
    KnownCapabilityGroup["SupportedElasticPoolEditions"] = "supportedElasticPoolEditions";
    /** SupportedManagedInstanceVersions */
    KnownCapabilityGroup["SupportedManagedInstanceVersions"] = "supportedManagedInstanceVersions";
    /** SupportedInstancePoolEditions */
    KnownCapabilityGroup["SupportedInstancePoolEditions"] = "supportedInstancePoolEditions";
    /** SupportedManagedInstanceEditions */
    KnownCapabilityGroup["SupportedManagedInstanceEditions"] = "supportedManagedInstanceEditions";
})(KnownCapabilityGroup || (KnownCapabilityGroup = {}));
/** Known values of {@link MaxSizeUnit} that the service accepts. */
export var KnownMaxSizeUnit;
(function (KnownMaxSizeUnit) {
    /** Megabytes */
    KnownMaxSizeUnit["Megabytes"] = "Megabytes";
    /** Gigabytes */
    KnownMaxSizeUnit["Gigabytes"] = "Gigabytes";
    /** Terabytes */
    KnownMaxSizeUnit["Terabytes"] = "Terabytes";
    /** Petabytes */
    KnownMaxSizeUnit["Petabytes"] = "Petabytes";
})(KnownMaxSizeUnit || (KnownMaxSizeUnit = {}));
/** Known values of {@link LogSizeUnit} that the service accepts. */
export var KnownLogSizeUnit;
(function (KnownLogSizeUnit) {
    /** Megabytes */
    KnownLogSizeUnit["Megabytes"] = "Megabytes";
    /** Gigabytes */
    KnownLogSizeUnit["Gigabytes"] = "Gigabytes";
    /** Terabytes */
    KnownLogSizeUnit["Terabytes"] = "Terabytes";
    /** Petabytes */
    KnownLogSizeUnit["Petabytes"] = "Petabytes";
    /** Percent */
    KnownLogSizeUnit["Percent"] = "Percent";
})(KnownLogSizeUnit || (KnownLogSizeUnit = {}));
/** Known values of {@link PerformanceLevelUnit} that the service accepts. */
export var KnownPerformanceLevelUnit;
(function (KnownPerformanceLevelUnit) {
    /** DTU */
    KnownPerformanceLevelUnit["DTU"] = "DTU";
    /** VCores */
    KnownPerformanceLevelUnit["VCores"] = "VCores";
})(KnownPerformanceLevelUnit || (KnownPerformanceLevelUnit = {}));
/** Known values of {@link PauseDelayTimeUnit} that the service accepts. */
export var KnownPauseDelayTimeUnit;
(function (KnownPauseDelayTimeUnit) {
    /** Minutes */
    KnownPauseDelayTimeUnit["Minutes"] = "Minutes";
})(KnownPauseDelayTimeUnit || (KnownPauseDelayTimeUnit = {}));
/** Known values of {@link StorageCapabilityStorageAccountType} that the service accepts. */
export var KnownStorageCapabilityStorageAccountType;
(function (KnownStorageCapabilityStorageAccountType) {
    /** GRS */
    KnownStorageCapabilityStorageAccountType["GRS"] = "GRS";
    /** LRS */
    KnownStorageCapabilityStorageAccountType["LRS"] = "LRS";
    /** ZRS */
    KnownStorageCapabilityStorageAccountType["ZRS"] = "ZRS";
})(KnownStorageCapabilityStorageAccountType || (KnownStorageCapabilityStorageAccountType = {}));
/** Known values of {@link LongTermRetentionPolicyName} that the service accepts. */
export var KnownLongTermRetentionPolicyName;
(function (KnownLongTermRetentionPolicyName) {
    /** Default */
    KnownLongTermRetentionPolicyName["Default"] = "default";
})(KnownLongTermRetentionPolicyName || (KnownLongTermRetentionPolicyName = {}));
/** Known values of {@link DayOfWeek} that the service accepts. */
export var KnownDayOfWeek;
(function (KnownDayOfWeek) {
    /** Sunday */
    KnownDayOfWeek["Sunday"] = "Sunday";
    /** Monday */
    KnownDayOfWeek["Monday"] = "Monday";
    /** Tuesday */
    KnownDayOfWeek["Tuesday"] = "Tuesday";
    /** Wednesday */
    KnownDayOfWeek["Wednesday"] = "Wednesday";
    /** Thursday */
    KnownDayOfWeek["Thursday"] = "Thursday";
    /** Friday */
    KnownDayOfWeek["Friday"] = "Friday";
    /** Saturday */
    KnownDayOfWeek["Saturday"] = "Saturday";
})(KnownDayOfWeek || (KnownDayOfWeek = {}));
/** Known values of {@link ManagedShortTermRetentionPolicyName} that the service accepts. */
export var KnownManagedShortTermRetentionPolicyName;
(function (KnownManagedShortTermRetentionPolicyName) {
    /** Default */
    KnownManagedShortTermRetentionPolicyName["Default"] = "default";
})(KnownManagedShortTermRetentionPolicyName || (KnownManagedShortTermRetentionPolicyName = {}));
/** Known values of {@link QueryTimeGrainType} that the service accepts. */
export var KnownQueryTimeGrainType;
(function (KnownQueryTimeGrainType) {
    /** PT1H */
    KnownQueryTimeGrainType["PT1H"] = "PT1H";
    /** P1D */
    KnownQueryTimeGrainType["P1D"] = "P1D";
})(KnownQueryTimeGrainType || (KnownQueryTimeGrainType = {}));
/** Known values of {@link QueryMetricUnitType} that the service accepts. */
export var KnownQueryMetricUnitType;
(function (KnownQueryMetricUnitType) {
    /** Percentage */
    KnownQueryMetricUnitType["Percentage"] = "percentage";
    /** KB */
    KnownQueryMetricUnitType["KB"] = "KB";
    /** Microseconds */
    KnownQueryMetricUnitType["Microseconds"] = "microseconds";
    /** Count */
    KnownQueryMetricUnitType["Count"] = "count";
})(KnownQueryMetricUnitType || (KnownQueryMetricUnitType = {}));
/** Known values of {@link RestoreDetailsName} that the service accepts. */
export var KnownRestoreDetailsName;
(function (KnownRestoreDetailsName) {
    /** Default */
    KnownRestoreDetailsName["Default"] = "Default";
})(KnownRestoreDetailsName || (KnownRestoreDetailsName = {}));
/** Known values of {@link ManagedDatabaseStatus} that the service accepts. */
export var KnownManagedDatabaseStatus;
(function (KnownManagedDatabaseStatus) {
    /** Online */
    KnownManagedDatabaseStatus["Online"] = "Online";
    /** Offline */
    KnownManagedDatabaseStatus["Offline"] = "Offline";
    /** Shutdown */
    KnownManagedDatabaseStatus["Shutdown"] = "Shutdown";
    /** Creating */
    KnownManagedDatabaseStatus["Creating"] = "Creating";
    /** Inaccessible */
    KnownManagedDatabaseStatus["Inaccessible"] = "Inaccessible";
    /** Restoring */
    KnownManagedDatabaseStatus["Restoring"] = "Restoring";
    /** Updating */
    KnownManagedDatabaseStatus["Updating"] = "Updating";
})(KnownManagedDatabaseStatus || (KnownManagedDatabaseStatus = {}));
/** Known values of {@link CatalogCollationType} that the service accepts. */
export var KnownCatalogCollationType;
(function (KnownCatalogCollationType) {
    /** DatabaseDefault */
    KnownCatalogCollationType["DatabaseDefault"] = "DATABASE_DEFAULT";
    /** SQLLatin1GeneralCP1CIAS */
    KnownCatalogCollationType["SQLLatin1GeneralCP1CIAS"] = "SQL_Latin1_General_CP1_CI_AS";
})(KnownCatalogCollationType || (KnownCatalogCollationType = {}));
/** Known values of {@link ManagedDatabaseCreateMode} that the service accepts. */
export var KnownManagedDatabaseCreateMode;
(function (KnownManagedDatabaseCreateMode) {
    /** Default */
    KnownManagedDatabaseCreateMode["Default"] = "Default";
    /** RestoreExternalBackup */
    KnownManagedDatabaseCreateMode["RestoreExternalBackup"] = "RestoreExternalBackup";
    /** PointInTimeRestore */
    KnownManagedDatabaseCreateMode["PointInTimeRestore"] = "PointInTimeRestore";
    /** Recovery */
    KnownManagedDatabaseCreateMode["Recovery"] = "Recovery";
    /** RestoreLongTermRetentionBackup */
    KnownManagedDatabaseCreateMode["RestoreLongTermRetentionBackup"] = "RestoreLongTermRetentionBackup";
})(KnownManagedDatabaseCreateMode || (KnownManagedDatabaseCreateMode = {}));
/** Known values of {@link TransparentDataEncryptionName} that the service accepts. */
export var KnownTransparentDataEncryptionName;
(function (KnownTransparentDataEncryptionName) {
    /** Current */
    KnownTransparentDataEncryptionName["Current"] = "current";
})(KnownTransparentDataEncryptionName || (KnownTransparentDataEncryptionName = {}));
/** Known values of {@link ManagedInstanceAdministratorType} that the service accepts. */
export var KnownManagedInstanceAdministratorType;
(function (KnownManagedInstanceAdministratorType) {
    /** ActiveDirectory */
    KnownManagedInstanceAdministratorType["ActiveDirectory"] = "ActiveDirectory";
})(KnownManagedInstanceAdministratorType || (KnownManagedInstanceAdministratorType = {}));
/** Known values of {@link AdministratorName} that the service accepts. */
export var KnownAdministratorName;
(function (KnownAdministratorName) {
    /** ActiveDirectory */
    KnownAdministratorName["ActiveDirectory"] = "ActiveDirectory";
})(KnownAdministratorName || (KnownAdministratorName = {}));
/** Known values of {@link AuthenticationName} that the service accepts. */
export var KnownAuthenticationName;
(function (KnownAuthenticationName) {
    /** Default */
    KnownAuthenticationName["Default"] = "Default";
})(KnownAuthenticationName || (KnownAuthenticationName = {}));
/** Known values of {@link ManagedInstanceLongTermRetentionPolicyName} that the service accepts. */
export var KnownManagedInstanceLongTermRetentionPolicyName;
(function (KnownManagedInstanceLongTermRetentionPolicyName) {
    /** Default */
    KnownManagedInstanceLongTermRetentionPolicyName["Default"] = "default";
})(KnownManagedInstanceLongTermRetentionPolicyName || (KnownManagedInstanceLongTermRetentionPolicyName = {}));
/** Known values of {@link ManagementOperationState} that the service accepts. */
export var KnownManagementOperationState;
(function (KnownManagementOperationState) {
    /** Pending */
    KnownManagementOperationState["Pending"] = "Pending";
    /** InProgress */
    KnownManagementOperationState["InProgress"] = "InProgress";
    /** Succeeded */
    KnownManagementOperationState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownManagementOperationState["Failed"] = "Failed";
    /** CancelInProgress */
    KnownManagementOperationState["CancelInProgress"] = "CancelInProgress";
    /** Cancelled */
    KnownManagementOperationState["Cancelled"] = "Cancelled";
})(KnownManagementOperationState || (KnownManagementOperationState = {}));
/** Known values of {@link UpsertManagedServerOperationStepStatus} that the service accepts. */
export var KnownUpsertManagedServerOperationStepStatus;
(function (KnownUpsertManagedServerOperationStepStatus) {
    /** NotStarted */
    KnownUpsertManagedServerOperationStepStatus["NotStarted"] = "NotStarted";
    /** InProgress */
    KnownUpsertManagedServerOperationStepStatus["InProgress"] = "InProgress";
    /** SlowedDown */
    KnownUpsertManagedServerOperationStepStatus["SlowedDown"] = "SlowedDown";
    /** Completed */
    KnownUpsertManagedServerOperationStepStatus["Completed"] = "Completed";
    /** Failed */
    KnownUpsertManagedServerOperationStepStatus["Failed"] = "Failed";
    /** Canceled */
    KnownUpsertManagedServerOperationStepStatus["Canceled"] = "Canceled";
})(KnownUpsertManagedServerOperationStepStatus || (KnownUpsertManagedServerOperationStepStatus = {}));
/** Known values of {@link OperationOrigin} that the service accepts. */
export var KnownOperationOrigin;
(function (KnownOperationOrigin) {
    /** User */
    KnownOperationOrigin["User"] = "user";
    /** System */
    KnownOperationOrigin["System"] = "system";
})(KnownOperationOrigin || (KnownOperationOrigin = {}));
/** Known values of {@link PrivateLinkServiceConnectionStateStatus} that the service accepts. */
export var KnownPrivateLinkServiceConnectionStateStatus;
(function (KnownPrivateLinkServiceConnectionStateStatus) {
    /** Approved */
    KnownPrivateLinkServiceConnectionStateStatus["Approved"] = "Approved";
    /** Pending */
    KnownPrivateLinkServiceConnectionStateStatus["Pending"] = "Pending";
    /** Rejected */
    KnownPrivateLinkServiceConnectionStateStatus["Rejected"] = "Rejected";
    /** Disconnected */
    KnownPrivateLinkServiceConnectionStateStatus["Disconnected"] = "Disconnected";
})(KnownPrivateLinkServiceConnectionStateStatus || (KnownPrivateLinkServiceConnectionStateStatus = {}));
/** Known values of {@link PrivateLinkServiceConnectionStateActionsRequire} that the service accepts. */
export var KnownPrivateLinkServiceConnectionStateActionsRequire;
(function (KnownPrivateLinkServiceConnectionStateActionsRequire) {
    /** None */
    KnownPrivateLinkServiceConnectionStateActionsRequire["None"] = "None";
})(KnownPrivateLinkServiceConnectionStateActionsRequire || (KnownPrivateLinkServiceConnectionStateActionsRequire = {}));
/** Known values of {@link PrivateEndpointProvisioningState} that the service accepts. */
export var KnownPrivateEndpointProvisioningState;
(function (KnownPrivateEndpointProvisioningState) {
    /** Approving */
    KnownPrivateEndpointProvisioningState["Approving"] = "Approving";
    /** Ready */
    KnownPrivateEndpointProvisioningState["Ready"] = "Ready";
    /** Dropping */
    KnownPrivateEndpointProvisioningState["Dropping"] = "Dropping";
    /** Failed */
    KnownPrivateEndpointProvisioningState["Failed"] = "Failed";
    /** Rejecting */
    KnownPrivateEndpointProvisioningState["Rejecting"] = "Rejecting";
})(KnownPrivateEndpointProvisioningState || (KnownPrivateEndpointProvisioningState = {}));
/** Known values of {@link AdministratorType} that the service accepts. */
export var KnownAdministratorType;
(function (KnownAdministratorType) {
    /** ActiveDirectory */
    KnownAdministratorType["ActiveDirectory"] = "ActiveDirectory";
})(KnownAdministratorType || (KnownAdministratorType = {}));
/** Known values of {@link ServerTrustGroupPropertiesTrustScopesItem} that the service accepts. */
export var KnownServerTrustGroupPropertiesTrustScopesItem;
(function (KnownServerTrustGroupPropertiesTrustScopesItem) {
    /** GlobalTransactions */
    KnownServerTrustGroupPropertiesTrustScopesItem["GlobalTransactions"] = "GlobalTransactions";
    /** ServiceBroker */
    KnownServerTrustGroupPropertiesTrustScopesItem["ServiceBroker"] = "ServiceBroker";
})(KnownServerTrustGroupPropertiesTrustScopesItem || (KnownServerTrustGroupPropertiesTrustScopesItem = {}));
/** Known values of {@link SqlAgentConfigurationPropertiesState} that the service accepts. */
export var KnownSqlAgentConfigurationPropertiesState;
(function (KnownSqlAgentConfigurationPropertiesState) {
    /** Enabled */
    KnownSqlAgentConfigurationPropertiesState["Enabled"] = "Enabled";
    /** Disabled */
    KnownSqlAgentConfigurationPropertiesState["Disabled"] = "Disabled";
})(KnownSqlAgentConfigurationPropertiesState || (KnownSqlAgentConfigurationPropertiesState = {}));
/** Known values of {@link SyncAgentState} that the service accepts. */
export var KnownSyncAgentState;
(function (KnownSyncAgentState) {
    /** Online */
    KnownSyncAgentState["Online"] = "Online";
    /** Offline */
    KnownSyncAgentState["Offline"] = "Offline";
    /** NeverConnected */
    KnownSyncAgentState["NeverConnected"] = "NeverConnected";
})(KnownSyncAgentState || (KnownSyncAgentState = {}));
/** Known values of {@link SyncMemberDbType} that the service accepts. */
export var KnownSyncMemberDbType;
(function (KnownSyncMemberDbType) {
    /** AzureSqlDatabase */
    KnownSyncMemberDbType["AzureSqlDatabase"] = "AzureSqlDatabase";
    /** SqlServerDatabase */
    KnownSyncMemberDbType["SqlServerDatabase"] = "SqlServerDatabase";
})(KnownSyncMemberDbType || (KnownSyncMemberDbType = {}));
/** Known values of {@link Enum60} that the service accepts. */
export var KnownEnum60;
(function (KnownEnum60) {
    /** All */
    KnownEnum60["All"] = "All";
    /** Error */
    KnownEnum60["Error"] = "Error";
    /** Warning */
    KnownEnum60["Warning"] = "Warning";
    /** Success */
    KnownEnum60["Success"] = "Success";
})(KnownEnum60 || (KnownEnum60 = {}));
/** Known values of {@link SyncGroupLogType} that the service accepts. */
export var KnownSyncGroupLogType;
(function (KnownSyncGroupLogType) {
    /** All */
    KnownSyncGroupLogType["All"] = "All";
    /** Error */
    KnownSyncGroupLogType["Error"] = "Error";
    /** Warning */
    KnownSyncGroupLogType["Warning"] = "Warning";
    /** Success */
    KnownSyncGroupLogType["Success"] = "Success";
})(KnownSyncGroupLogType || (KnownSyncGroupLogType = {}));
/** Known values of {@link SyncConflictResolutionPolicy} that the service accepts. */
export var KnownSyncConflictResolutionPolicy;
(function (KnownSyncConflictResolutionPolicy) {
    /** HubWin */
    KnownSyncConflictResolutionPolicy["HubWin"] = "HubWin";
    /** MemberWin */
    KnownSyncConflictResolutionPolicy["MemberWin"] = "MemberWin";
})(KnownSyncConflictResolutionPolicy || (KnownSyncConflictResolutionPolicy = {}));
/** Known values of {@link SyncGroupState} that the service accepts. */
export var KnownSyncGroupState;
(function (KnownSyncGroupState) {
    /** NotReady */
    KnownSyncGroupState["NotReady"] = "NotReady";
    /** Error */
    KnownSyncGroupState["Error"] = "Error";
    /** Warning */
    KnownSyncGroupState["Warning"] = "Warning";
    /** Progressing */
    KnownSyncGroupState["Progressing"] = "Progressing";
    /** Good */
    KnownSyncGroupState["Good"] = "Good";
})(KnownSyncGroupState || (KnownSyncGroupState = {}));
/** Known values of {@link SyncDirection} that the service accepts. */
export var KnownSyncDirection;
(function (KnownSyncDirection) {
    /** Bidirectional */
    KnownSyncDirection["Bidirectional"] = "Bidirectional";
    /** OneWayMemberToHub */
    KnownSyncDirection["OneWayMemberToHub"] = "OneWayMemberToHub";
    /** OneWayHubToMember */
    KnownSyncDirection["OneWayHubToMember"] = "OneWayHubToMember";
})(KnownSyncDirection || (KnownSyncDirection = {}));
/** Known values of {@link SyncMemberState} that the service accepts. */
export var KnownSyncMemberState;
(function (KnownSyncMemberState) {
    /** SyncInProgress */
    KnownSyncMemberState["SyncInProgress"] = "SyncInProgress";
    /** SyncSucceeded */
    KnownSyncMemberState["SyncSucceeded"] = "SyncSucceeded";
    /** SyncFailed */
    KnownSyncMemberState["SyncFailed"] = "SyncFailed";
    /** DisabledTombstoneCleanup */
    KnownSyncMemberState["DisabledTombstoneCleanup"] = "DisabledTombstoneCleanup";
    /** DisabledBackupRestore */
    KnownSyncMemberState["DisabledBackupRestore"] = "DisabledBackupRestore";
    /** SyncSucceededWithWarnings */
    KnownSyncMemberState["SyncSucceededWithWarnings"] = "SyncSucceededWithWarnings";
    /** SyncCancelling */
    KnownSyncMemberState["SyncCancelling"] = "SyncCancelling";
    /** SyncCancelled */
    KnownSyncMemberState["SyncCancelled"] = "SyncCancelled";
    /** UnProvisioned */
    KnownSyncMemberState["UnProvisioned"] = "UnProvisioned";
    /** Provisioning */
    KnownSyncMemberState["Provisioning"] = "Provisioning";
    /** Provisioned */
    KnownSyncMemberState["Provisioned"] = "Provisioned";
    /** ProvisionFailed */
    KnownSyncMemberState["ProvisionFailed"] = "ProvisionFailed";
    /** DeProvisioning */
    KnownSyncMemberState["DeProvisioning"] = "DeProvisioning";
    /** DeProvisioned */
    KnownSyncMemberState["DeProvisioned"] = "DeProvisioned";
    /** DeProvisionFailed */
    KnownSyncMemberState["DeProvisionFailed"] = "DeProvisionFailed";
    /** Reprovisioning */
    KnownSyncMemberState["Reprovisioning"] = "Reprovisioning";
    /** ReprovisionFailed */
    KnownSyncMemberState["ReprovisionFailed"] = "ReprovisionFailed";
    /** UnReprovisioned */
    KnownSyncMemberState["UnReprovisioned"] = "UnReprovisioned";
})(KnownSyncMemberState || (KnownSyncMemberState = {}));
/** Known values of {@link DnsRefreshConfigurationPropertiesStatus} that the service accepts. */
export var KnownDnsRefreshConfigurationPropertiesStatus;
(function (KnownDnsRefreshConfigurationPropertiesStatus) {
    /** Succeeded */
    KnownDnsRefreshConfigurationPropertiesStatus["Succeeded"] = "Succeeded";
    /** Failed */
    KnownDnsRefreshConfigurationPropertiesStatus["Failed"] = "Failed";
})(KnownDnsRefreshConfigurationPropertiesStatus || (KnownDnsRefreshConfigurationPropertiesStatus = {}));
/** Known values of {@link VirtualNetworkRuleState} that the service accepts. */
export var KnownVirtualNetworkRuleState;
(function (KnownVirtualNetworkRuleState) {
    /** Initializing */
    KnownVirtualNetworkRuleState["Initializing"] = "Initializing";
    /** InProgress */
    KnownVirtualNetworkRuleState["InProgress"] = "InProgress";
    /** Ready */
    KnownVirtualNetworkRuleState["Ready"] = "Ready";
    /** Failed */
    KnownVirtualNetworkRuleState["Failed"] = "Failed";
    /** Deleting */
    KnownVirtualNetworkRuleState["Deleting"] = "Deleting";
    /** Unknown */
    KnownVirtualNetworkRuleState["Unknown"] = "Unknown";
})(KnownVirtualNetworkRuleState || (KnownVirtualNetworkRuleState = {}));
/** Known values of {@link ShortTermRetentionPolicyName} that the service accepts. */
export var KnownShortTermRetentionPolicyName;
(function (KnownShortTermRetentionPolicyName) {
    /** Default */
    KnownShortTermRetentionPolicyName["Default"] = "default";
})(KnownShortTermRetentionPolicyName || (KnownShortTermRetentionPolicyName = {}));
/** Known values of {@link DiffBackupIntervalInHours} that the service accepts. */
export var KnownDiffBackupIntervalInHours;
(function (KnownDiffBackupIntervalInHours) {
    /** Twelve */
    KnownDiffBackupIntervalInHours[KnownDiffBackupIntervalInHours["Twelve"] = 12] = "Twelve";
    /** TwentyFour */
    KnownDiffBackupIntervalInHours[KnownDiffBackupIntervalInHours["TwentyFour"] = 24] = "TwentyFour";
})(KnownDiffBackupIntervalInHours || (KnownDiffBackupIntervalInHours = {}));
/** Known values of {@link OperationMode} that the service accepts. */
export var KnownOperationMode;
(function (KnownOperationMode) {
    /** PolybaseImport */
    KnownOperationMode["PolybaseImport"] = "PolybaseImport";
})(KnownOperationMode || (KnownOperationMode = {}));
/** Known values of {@link StorageKeyType} that the service accepts. */
export var KnownStorageKeyType;
(function (KnownStorageKeyType) {
    /** SharedAccessKey */
    KnownStorageKeyType["SharedAccessKey"] = "SharedAccessKey";
    /** StorageAccessKey */
    KnownStorageKeyType["StorageAccessKey"] = "StorageAccessKey";
})(KnownStorageKeyType || (KnownStorageKeyType = {}));
/** Known values of {@link LedgerDigestUploadsName} that the service accepts. */
export var KnownLedgerDigestUploadsName;
(function (KnownLedgerDigestUploadsName) {
    /** Current */
    KnownLedgerDigestUploadsName["Current"] = "current";
})(KnownLedgerDigestUploadsName || (KnownLedgerDigestUploadsName = {}));
/** Known values of {@link ReplicationState} that the service accepts. */
export var KnownReplicationState;
(function (KnownReplicationState) {
    /** Pending */
    KnownReplicationState["Pending"] = "PENDING";
    /** Seeding */
    KnownReplicationState["Seeding"] = "SEEDING";
    /** CatchUP */
    KnownReplicationState["CatchUP"] = "CATCH_UP";
    /** Suspended */
    KnownReplicationState["Suspended"] = "SUSPENDED";
})(KnownReplicationState || (KnownReplicationState = {}));
/** Known values of {@link ReplicationLinkType} that the service accepts. */
export var KnownReplicationLinkType;
(function (KnownReplicationLinkType) {
    /** GEO */
    KnownReplicationLinkType["GEO"] = "GEO";
    /** Named */
    KnownReplicationLinkType["Named"] = "NAMED";
})(KnownReplicationLinkType || (KnownReplicationLinkType = {}));
/** Known values of {@link IdentityType} that the service accepts. */
export var KnownIdentityType;
(function (KnownIdentityType) {
    /** None */
    KnownIdentityType["None"] = "None";
    /** SystemAssigned */
    KnownIdentityType["SystemAssigned"] = "SystemAssigned";
    /** UserAssigned */
    KnownIdentityType["UserAssigned"] = "UserAssigned";
    /** SystemAssignedUserAssigned */
    KnownIdentityType["SystemAssignedUserAssigned"] = "SystemAssigned,UserAssigned";
})(KnownIdentityType || (KnownIdentityType = {}));
/** Known values of {@link ServerNetworkAccessFlag} that the service accepts. */
export var KnownServerNetworkAccessFlag;
(function (KnownServerNetworkAccessFlag) {
    /** Enabled */
    KnownServerNetworkAccessFlag["Enabled"] = "Enabled";
    /** Disabled */
    KnownServerNetworkAccessFlag["Disabled"] = "Disabled";
})(KnownServerNetworkAccessFlag || (KnownServerNetworkAccessFlag = {}));
/** Known values of {@link ServerWorkspaceFeature} that the service accepts. */
export var KnownServerWorkspaceFeature;
(function (KnownServerWorkspaceFeature) {
    /** Connected */
    KnownServerWorkspaceFeature["Connected"] = "Connected";
    /** Disconnected */
    KnownServerWorkspaceFeature["Disconnected"] = "Disconnected";
})(KnownServerWorkspaceFeature || (KnownServerWorkspaceFeature = {}));
/** Known values of {@link PrincipalType} that the service accepts. */
export var KnownPrincipalType;
(function (KnownPrincipalType) {
    /** User */
    KnownPrincipalType["User"] = "User";
    /** Group */
    KnownPrincipalType["Group"] = "Group";
    /** Application */
    KnownPrincipalType["Application"] = "Application";
})(KnownPrincipalType || (KnownPrincipalType = {}));
/** Known values of {@link DatabaseIdentityType} that the service accepts. */
export var KnownDatabaseIdentityType;
(function (KnownDatabaseIdentityType) {
    /** None */
    KnownDatabaseIdentityType["None"] = "None";
    /** UserAssigned */
    KnownDatabaseIdentityType["UserAssigned"] = "UserAssigned";
})(KnownDatabaseIdentityType || (KnownDatabaseIdentityType = {}));
/** Known values of {@link CreateMode} that the service accepts. */
export var KnownCreateMode;
(function (KnownCreateMode) {
    /** Default */
    KnownCreateMode["Default"] = "Default";
    /** Copy */
    KnownCreateMode["Copy"] = "Copy";
    /** Secondary */
    KnownCreateMode["Secondary"] = "Secondary";
    /** PointInTimeRestore */
    KnownCreateMode["PointInTimeRestore"] = "PointInTimeRestore";
    /** Restore */
    KnownCreateMode["Restore"] = "Restore";
    /** Recovery */
    KnownCreateMode["Recovery"] = "Recovery";
    /** RestoreExternalBackup */
    KnownCreateMode["RestoreExternalBackup"] = "RestoreExternalBackup";
    /** RestoreExternalBackupSecondary */
    KnownCreateMode["RestoreExternalBackupSecondary"] = "RestoreExternalBackupSecondary";
    /** RestoreLongTermRetentionBackup */
    KnownCreateMode["RestoreLongTermRetentionBackup"] = "RestoreLongTermRetentionBackup";
    /** OnlineSecondary */
    KnownCreateMode["OnlineSecondary"] = "OnlineSecondary";
})(KnownCreateMode || (KnownCreateMode = {}));
/** Known values of {@link SampleName} that the service accepts. */
export var KnownSampleName;
(function (KnownSampleName) {
    /** AdventureWorksLT */
    KnownSampleName["AdventureWorksLT"] = "AdventureWorksLT";
    /** WideWorldImportersStd */
    KnownSampleName["WideWorldImportersStd"] = "WideWorldImportersStd";
    /** WideWorldImportersFull */
    KnownSampleName["WideWorldImportersFull"] = "WideWorldImportersFull";
})(KnownSampleName || (KnownSampleName = {}));
/** Known values of {@link DatabaseStatus} that the service accepts. */
export var KnownDatabaseStatus;
(function (KnownDatabaseStatus) {
    /** Online */
    KnownDatabaseStatus["Online"] = "Online";
    /** Restoring */
    KnownDatabaseStatus["Restoring"] = "Restoring";
    /** RecoveryPending */
    KnownDatabaseStatus["RecoveryPending"] = "RecoveryPending";
    /** Recovering */
    KnownDatabaseStatus["Recovering"] = "Recovering";
    /** Suspect */
    KnownDatabaseStatus["Suspect"] = "Suspect";
    /** Offline */
    KnownDatabaseStatus["Offline"] = "Offline";
    /** Standby */
    KnownDatabaseStatus["Standby"] = "Standby";
    /** Shutdown */
    KnownDatabaseStatus["Shutdown"] = "Shutdown";
    /** EmergencyMode */
    KnownDatabaseStatus["EmergencyMode"] = "EmergencyMode";
    /** AutoClosed */
    KnownDatabaseStatus["AutoClosed"] = "AutoClosed";
    /** Copying */
    KnownDatabaseStatus["Copying"] = "Copying";
    /** Creating */
    KnownDatabaseStatus["Creating"] = "Creating";
    /** Inaccessible */
    KnownDatabaseStatus["Inaccessible"] = "Inaccessible";
    /** OfflineSecondary */
    KnownDatabaseStatus["OfflineSecondary"] = "OfflineSecondary";
    /** Pausing */
    KnownDatabaseStatus["Pausing"] = "Pausing";
    /** Paused */
    KnownDatabaseStatus["Paused"] = "Paused";
    /** Resuming */
    KnownDatabaseStatus["Resuming"] = "Resuming";
    /** Scaling */
    KnownDatabaseStatus["Scaling"] = "Scaling";
    /** OfflineChangingDwPerformanceTiers */
    KnownDatabaseStatus["OfflineChangingDwPerformanceTiers"] = "OfflineChangingDwPerformanceTiers";
    /** OnlineChangingDwPerformanceTiers */
    KnownDatabaseStatus["OnlineChangingDwPerformanceTiers"] = "OnlineChangingDwPerformanceTiers";
    /** Disabled */
    KnownDatabaseStatus["Disabled"] = "Disabled";
    /** Stopping */
    KnownDatabaseStatus["Stopping"] = "Stopping";
    /** Stopped */
    KnownDatabaseStatus["Stopped"] = "Stopped";
    /** Starting */
    KnownDatabaseStatus["Starting"] = "Starting";
})(KnownDatabaseStatus || (KnownDatabaseStatus = {}));
/** Known values of {@link DatabaseLicenseType} that the service accepts. */
export var KnownDatabaseLicenseType;
(function (KnownDatabaseLicenseType) {
    /** LicenseIncluded */
    KnownDatabaseLicenseType["LicenseIncluded"] = "LicenseIncluded";
    /** BasePrice */
    KnownDatabaseLicenseType["BasePrice"] = "BasePrice";
})(KnownDatabaseLicenseType || (KnownDatabaseLicenseType = {}));
/** Known values of {@link DatabaseReadScale} that the service accepts. */
export var KnownDatabaseReadScale;
(function (KnownDatabaseReadScale) {
    /** Enabled */
    KnownDatabaseReadScale["Enabled"] = "Enabled";
    /** Disabled */
    KnownDatabaseReadScale["Disabled"] = "Disabled";
})(KnownDatabaseReadScale || (KnownDatabaseReadScale = {}));
/** Known values of {@link SecondaryType} that the service accepts. */
export var KnownSecondaryType;
(function (KnownSecondaryType) {
    /** Geo */
    KnownSecondaryType["Geo"] = "Geo";
    /** Named */
    KnownSecondaryType["Named"] = "Named";
})(KnownSecondaryType || (KnownSecondaryType = {}));
/** Known values of {@link BackupStorageRedundancy} that the service accepts. */
export var KnownBackupStorageRedundancy;
(function (KnownBackupStorageRedundancy) {
    /** Geo */
    KnownBackupStorageRedundancy["Geo"] = "Geo";
    /** Local */
    KnownBackupStorageRedundancy["Local"] = "Local";
    /** Zone */
    KnownBackupStorageRedundancy["Zone"] = "Zone";
    /** GeoZone */
    KnownBackupStorageRedundancy["GeoZone"] = "GeoZone";
})(KnownBackupStorageRedundancy || (KnownBackupStorageRedundancy = {}));
/** Known values of {@link ReplicaType} that the service accepts. */
export var KnownReplicaType;
(function (KnownReplicaType) {
    /** Primary */
    KnownReplicaType["Primary"] = "Primary";
    /** ReadableSecondary */
    KnownReplicaType["ReadableSecondary"] = "ReadableSecondary";
})(KnownReplicaType || (KnownReplicaType = {}));
/** Known values of {@link DatabaseState} that the service accepts. */
export var KnownDatabaseState;
(function (KnownDatabaseState) {
    /** All */
    KnownDatabaseState["All"] = "All";
    /** Live */
    KnownDatabaseState["Live"] = "Live";
    /** Deleted */
    KnownDatabaseState["Deleted"] = "Deleted";
})(KnownDatabaseState || (KnownDatabaseState = {}));
/** Known values of {@link ManagedInstancePropertiesProvisioningState} that the service accepts. */
export var KnownManagedInstancePropertiesProvisioningState;
(function (KnownManagedInstancePropertiesProvisioningState) {
    /** Creating */
    KnownManagedInstancePropertiesProvisioningState["Creating"] = "Creating";
    /** Deleting */
    KnownManagedInstancePropertiesProvisioningState["Deleting"] = "Deleting";
    /** Updating */
    KnownManagedInstancePropertiesProvisioningState["Updating"] = "Updating";
    /** Unknown */
    KnownManagedInstancePropertiesProvisioningState["Unknown"] = "Unknown";
    /** Succeeded */
    KnownManagedInstancePropertiesProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownManagedInstancePropertiesProvisioningState["Failed"] = "Failed";
    /** Accepted */
    KnownManagedInstancePropertiesProvisioningState["Accepted"] = "Accepted";
    /** Created */
    KnownManagedInstancePropertiesProvisioningState["Created"] = "Created";
    /** Deleted */
    KnownManagedInstancePropertiesProvisioningState["Deleted"] = "Deleted";
    /** Unrecognized */
    KnownManagedInstancePropertiesProvisioningState["Unrecognized"] = "Unrecognized";
    /** Running */
    KnownManagedInstancePropertiesProvisioningState["Running"] = "Running";
    /** Canceled */
    KnownManagedInstancePropertiesProvisioningState["Canceled"] = "Canceled";
    /** NotSpecified */
    KnownManagedInstancePropertiesProvisioningState["NotSpecified"] = "NotSpecified";
    /** Registering */
    KnownManagedInstancePropertiesProvisioningState["Registering"] = "Registering";
    /** TimedOut */
    KnownManagedInstancePropertiesProvisioningState["TimedOut"] = "TimedOut";
})(KnownManagedInstancePropertiesProvisioningState || (KnownManagedInstancePropertiesProvisioningState = {}));
/** Known values of {@link ManagedServerCreateMode} that the service accepts. */
export var KnownManagedServerCreateMode;
(function (KnownManagedServerCreateMode) {
    /** Default */
    KnownManagedServerCreateMode["Default"] = "Default";
    /** PointInTimeRestore */
    KnownManagedServerCreateMode["PointInTimeRestore"] = "PointInTimeRestore";
})(KnownManagedServerCreateMode || (KnownManagedServerCreateMode = {}));
/** Known values of {@link ManagedInstanceLicenseType} that the service accepts. */
export var KnownManagedInstanceLicenseType;
(function (KnownManagedInstanceLicenseType) {
    /** LicenseIncluded */
    KnownManagedInstanceLicenseType["LicenseIncluded"] = "LicenseIncluded";
    /** BasePrice */
    KnownManagedInstanceLicenseType["BasePrice"] = "BasePrice";
})(KnownManagedInstanceLicenseType || (KnownManagedInstanceLicenseType = {}));
/** Known values of {@link ManagedInstanceProxyOverride} that the service accepts. */
export var KnownManagedInstanceProxyOverride;
(function (KnownManagedInstanceProxyOverride) {
    /** Proxy */
    KnownManagedInstanceProxyOverride["Proxy"] = "Proxy";
    /** Redirect */
    KnownManagedInstanceProxyOverride["Redirect"] = "Redirect";
    /** Default */
    KnownManagedInstanceProxyOverride["Default"] = "Default";
})(KnownManagedInstanceProxyOverride || (KnownManagedInstanceProxyOverride = {}));
/** Known values of {@link ServicePrincipalType} that the service accepts. */
export var KnownServicePrincipalType;
(function (KnownServicePrincipalType) {
    /** None */
    KnownServicePrincipalType["None"] = "None";
    /** SystemAssigned */
    KnownServicePrincipalType["SystemAssigned"] = "SystemAssigned";
})(KnownServicePrincipalType || (KnownServicePrincipalType = {}));
/** Known values of {@link AggregationFunctionType} that the service accepts. */
export var KnownAggregationFunctionType;
(function (KnownAggregationFunctionType) {
    /** Avg */
    KnownAggregationFunctionType["Avg"] = "avg";
    /** Min */
    KnownAggregationFunctionType["Min"] = "min";
    /** Max */
    KnownAggregationFunctionType["Max"] = "max";
    /** Stdev */
    KnownAggregationFunctionType["Stdev"] = "stdev";
    /** Sum */
    KnownAggregationFunctionType["Sum"] = "sum";
})(KnownAggregationFunctionType || (KnownAggregationFunctionType = {}));
/** Known values of {@link MetricType} that the service accepts. */
export var KnownMetricType;
(function (KnownMetricType) {
    /** Cpu */
    KnownMetricType["Cpu"] = "cpu";
    /** Io */
    KnownMetricType["Io"] = "io";
    /** LogIo */
    KnownMetricType["LogIo"] = "logIo";
    /** Duration */
    KnownMetricType["Duration"] = "duration";
    /** Dtu */
    KnownMetricType["Dtu"] = "dtu";
})(KnownMetricType || (KnownMetricType = {}));
/** Known values of {@link ConnectionPolicyName} that the service accepts. */
export var KnownConnectionPolicyName;
(function (KnownConnectionPolicyName) {
    /** Default */
    KnownConnectionPolicyName["Default"] = "default";
})(KnownConnectionPolicyName || (KnownConnectionPolicyName = {}));
/** Known values of {@link ServerConnectionType} that the service accepts. */
export var KnownServerConnectionType;
(function (KnownServerConnectionType) {
    /** Default */
    KnownServerConnectionType["Default"] = "Default";
    /** Redirect */
    KnownServerConnectionType["Redirect"] = "Redirect";
    /** Proxy */
    KnownServerConnectionType["Proxy"] = "Proxy";
})(KnownServerConnectionType || (KnownServerConnectionType = {}));
/** Known values of {@link ServiceObjectiveName} that the service accepts. */
export var KnownServiceObjectiveName;
(function (KnownServiceObjectiveName) {
    /** System */
    KnownServiceObjectiveName["System"] = "System";
    /** System0 */
    KnownServiceObjectiveName["System0"] = "System0";
    /** System1 */
    KnownServiceObjectiveName["System1"] = "System1";
    /** System2 */
    KnownServiceObjectiveName["System2"] = "System2";
    /** System3 */
    KnownServiceObjectiveName["System3"] = "System3";
    /** System4 */
    KnownServiceObjectiveName["System4"] = "System4";
    /** System2L */
    KnownServiceObjectiveName["System2L"] = "System2L";
    /** System3L */
    KnownServiceObjectiveName["System3L"] = "System3L";
    /** System4L */
    KnownServiceObjectiveName["System4L"] = "System4L";
    /** Free */
    KnownServiceObjectiveName["Free"] = "Free";
    /** Basic */
    KnownServiceObjectiveName["Basic"] = "Basic";
    /** S0 */
    KnownServiceObjectiveName["S0"] = "S0";
    /** S1 */
    KnownServiceObjectiveName["S1"] = "S1";
    /** S2 */
    KnownServiceObjectiveName["S2"] = "S2";
    /** S3 */
    KnownServiceObjectiveName["S3"] = "S3";
    /** S4 */
    KnownServiceObjectiveName["S4"] = "S4";
    /** S6 */
    KnownServiceObjectiveName["S6"] = "S6";
    /** S7 */
    KnownServiceObjectiveName["S7"] = "S7";
    /** S9 */
    KnownServiceObjectiveName["S9"] = "S9";
    /** S12 */
    KnownServiceObjectiveName["S12"] = "S12";
    /** P1 */
    KnownServiceObjectiveName["P1"] = "P1";
    /** P2 */
    KnownServiceObjectiveName["P2"] = "P2";
    /** P3 */
    KnownServiceObjectiveName["P3"] = "P3";
    /** P4 */
    KnownServiceObjectiveName["P4"] = "P4";
    /** P6 */
    KnownServiceObjectiveName["P6"] = "P6";
    /** P11 */
    KnownServiceObjectiveName["P11"] = "P11";
    /** P15 */
    KnownServiceObjectiveName["P15"] = "P15";
    /** PRS1 */
    KnownServiceObjectiveName["PRS1"] = "PRS1";
    /** PRS2 */
    KnownServiceObjectiveName["PRS2"] = "PRS2";
    /** PRS4 */
    KnownServiceObjectiveName["PRS4"] = "PRS4";
    /** PRS6 */
    KnownServiceObjectiveName["PRS6"] = "PRS6";
    /** DW100 */
    KnownServiceObjectiveName["DW100"] = "DW100";
    /** DW200 */
    KnownServiceObjectiveName["DW200"] = "DW200";
    /** DW300 */
    KnownServiceObjectiveName["DW300"] = "DW300";
    /** DW400 */
    KnownServiceObjectiveName["DW400"] = "DW400";
    /** DW500 */
    KnownServiceObjectiveName["DW500"] = "DW500";
    /** DW600 */
    KnownServiceObjectiveName["DW600"] = "DW600";
    /** DW1000 */
    KnownServiceObjectiveName["DW1000"] = "DW1000";
    /** DW1200 */
    KnownServiceObjectiveName["DW1200"] = "DW1200";
    /** DW1000C */
    KnownServiceObjectiveName["DW1000C"] = "DW1000c";
    /** DW1500 */
    KnownServiceObjectiveName["DW1500"] = "DW1500";
    /** DW1500C */
    KnownServiceObjectiveName["DW1500C"] = "DW1500c";
    /** DW2000 */
    KnownServiceObjectiveName["DW2000"] = "DW2000";
    /** DW2000C */
    KnownServiceObjectiveName["DW2000C"] = "DW2000c";
    /** DW3000 */
    KnownServiceObjectiveName["DW3000"] = "DW3000";
    /** DW2500C */
    KnownServiceObjectiveName["DW2500C"] = "DW2500c";
    /** DW3000C */
    KnownServiceObjectiveName["DW3000C"] = "DW3000c";
    /** DW6000 */
    KnownServiceObjectiveName["DW6000"] = "DW6000";
    /** DW5000C */
    KnownServiceObjectiveName["DW5000C"] = "DW5000c";
    /** DW6000C */
    KnownServiceObjectiveName["DW6000C"] = "DW6000c";
    /** DW7500C */
    KnownServiceObjectiveName["DW7500C"] = "DW7500c";
    /** DW10000C */
    KnownServiceObjectiveName["DW10000C"] = "DW10000c";
    /** DW15000C */
    KnownServiceObjectiveName["DW15000C"] = "DW15000c";
    /** DW30000C */
    KnownServiceObjectiveName["DW30000C"] = "DW30000c";
    /** DS100 */
    KnownServiceObjectiveName["DS100"] = "DS100";
    /** DS200 */
    KnownServiceObjectiveName["DS200"] = "DS200";
    /** DS300 */
    KnownServiceObjectiveName["DS300"] = "DS300";
    /** DS400 */
    KnownServiceObjectiveName["DS400"] = "DS400";
    /** DS500 */
    KnownServiceObjectiveName["DS500"] = "DS500";
    /** DS600 */
    KnownServiceObjectiveName["DS600"] = "DS600";
    /** DS1000 */
    KnownServiceObjectiveName["DS1000"] = "DS1000";
    /** DS1200 */
    KnownServiceObjectiveName["DS1200"] = "DS1200";
    /** DS1500 */
    KnownServiceObjectiveName["DS1500"] = "DS1500";
    /** DS2000 */
    KnownServiceObjectiveName["DS2000"] = "DS2000";
    /** ElasticPool */
    KnownServiceObjectiveName["ElasticPool"] = "ElasticPool";
})(KnownServiceObjectiveName || (KnownServiceObjectiveName = {}));
