// @azure/arm-sql/dist-esm/src/models/parameters.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { DataMaskingPolicy as DataMaskingPolicyMapper, DataMaskingRule as DataMaskingRuleMapper, GeoBackupPolicy as GeoBackupPolicyMapper, Database as DatabaseMapper, DatabaseUpdate as DatabaseUpdateMapper, ResourceMoveDefinition as ResourceMoveDefinitionMapper, ImportExistingDatabaseDefinition as ImportExistingDatabaseDefinitionMapper, ExportDatabaseDefinition as ExportDatabaseDefinitionMapper, ElasticPool as ElasticPoolMapper, ElasticPoolUpdate as ElasticPoolUpdateMapper, UnlinkParameters as UnlinkParametersMapper, ServerCommunicationLink as ServerCommunicationLinkMapper, ExtendedDatabaseBlobAuditingPolicy as ExtendedDatabaseBlobAuditingPolicyMapper, ExtendedServerBlobAuditingPolicy as ExtendedServerBlobAuditingPolicyMapper, ServerBlobAuditingPolicy as ServerBlobAuditingPolicyMapper, DatabaseBlobAuditingPolicy as DatabaseBlobAuditingPolicyMapper, Advisor as AdvisorMapper, DatabaseAutomaticTuning as DatabaseAutomaticTuningMapper, RecommendedAction as RecommendedActionMapper, DatabaseSecurityAlertPolicy as DatabaseSecurityAlertPolicyMapper, DatabaseVulnerabilityAssessmentRuleBaseline as DatabaseVulnerabilityAssessmentRuleBaselineMapper, DatabaseVulnerabilityAssessment as DatabaseVulnerabilityAssessmentMapper, EncryptionProtector as EncryptionProtectorMapper, FailoverGroup as FailoverGroupMapper, FailoverGroupUpdate as FailoverGroupUpdateMapper, FirewallRule as FirewallRuleMapper, FirewallRuleList as FirewallRuleListMapper, InstanceFailoverGroup as InstanceFailoverGroupMapper, InstancePool as InstancePoolMapper, InstancePoolUpdate as InstancePoolUpdateMapper, JobAgent as JobAgentMapper, JobAgentUpdate as JobAgentUpdateMapper, JobCredential as JobCredentialMapper, Job as JobMapper, JobStep as JobStepMapper, JobTargetGroup as JobTargetGroupMapper, LongTermRetentionPolicy as LongTermRetentionPolicyMapper, MaintenanceWindows as MaintenanceWindowsMapper, ManagedBackupShortTermRetentionPolicy as ManagedBackupShortTermRetentionPolicyMapper, ManagedDatabase as ManagedDatabaseMapper, ManagedDatabaseUpdate as ManagedDatabaseUpdateMapper, CompleteDatabaseRestoreDefinition as CompleteDatabaseRestoreDefinitionMapper, ManagedDatabaseSecurityAlertPolicy as ManagedDatabaseSecurityAlertPolicyMapper, SensitivityLabel as SensitivityLabelMapper, SensitivityLabelUpdateList as SensitivityLabelUpdateListMapper, RecommendedSensitivityLabelUpdateList as RecommendedSensitivityLabelUpdateListMapper, ManagedTransparentDataEncryption as ManagedTransparentDataEncryptionMapper, ManagedInstanceAdministrator as ManagedInstanceAdministratorMapper, ManagedInstanceAzureADOnlyAuthentication as ManagedInstanceAzureADOnlyAuthenticationMapper, ManagedInstanceEncryptionProtector as ManagedInstanceEncryptionProtectorMapper, ManagedInstanceKey as ManagedInstanceKeyMapper, ManagedInstanceLongTermRetentionPolicy as ManagedInstanceLongTermRetentionPolicyMapper, ManagedInstancePrivateEndpointConnection as ManagedInstancePrivateEndpointConnectionMapper, TdeCertificate as TdeCertificateMapper, ManagedInstanceVulnerabilityAssessment as ManagedInstanceVulnerabilityAssessmentMapper, ManagedServerSecurityAlertPolicy as ManagedServerSecurityAlertPolicyMapper, PrivateEndpointConnection as PrivateEndpointConnectionMapper, CreateDatabaseRestorePointDefinition as CreateDatabaseRestorePointDefinitionMapper, ServerAutomaticTuning as ServerAutomaticTuningMapper, ServerAzureADAdministrator as ServerAzureADAdministratorMapper, ServerAzureADOnlyAuthentication as ServerAzureADOnlyAuthenticationMapper, ServerDevOpsAuditingSettings as ServerDevOpsAuditingSettingsMapper, ServerDnsAliasAcquisition as ServerDnsAliasAcquisitionMapper, ServerKey as ServerKeyMapper, ServerSecurityAlertPolicy as ServerSecurityAlertPolicyMapper, ServerTrustGroup as ServerTrustGroupMapper, ServerVulnerabilityAssessment as ServerVulnerabilityAssessmentMapper, SqlAgentConfiguration as SqlAgentConfigurationMapper, SyncAgent as SyncAgentMapper, SyncGroup as SyncGroupMapper, SyncMember as SyncMemberMapper, VirtualClusterUpdate as VirtualClusterUpdateMapper, VirtualNetworkRule as VirtualNetworkRuleMapper, WorkloadClassifier as WorkloadClassifierMapper, WorkloadGroup as WorkloadGroupMapper, LogicalDatabaseTransparentDataEncryption as LogicalDatabaseTransparentDataEncryptionMapper, BackupShortTermRetentionPolicy as BackupShortTermRetentionPolicyMapper, DatabaseExtensions as DatabaseExtensionsMapper, LedgerDigestUploads as LedgerDigestUploadsMapper, OutboundFirewallRule as OutboundFirewallRuleMapper, Server as ServerMapper, ServerUpdate as ServerUpdateMapper, ImportNewDatabaseDefinition as ImportNewDatabaseDefinitionMapper, CheckNameAvailabilityRequest as CheckNameAvailabilityRequestMapper, CopyLongTermRetentionBackupParameters as CopyLongTermRetentionBackupParametersMapper, UpdateLongTermRetentionBackupParameters as UpdateLongTermRetentionBackupParametersMapper, ManagedInstance as ManagedInstanceMapper, ManagedInstanceUpdate as ManagedInstanceUpdateMapper, ServerConnectionPolicy as ServerConnectionPolicyMapper } from "../models/mappers";
export const accept = {
    parameterPath: "accept",
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Accept",
        type: {
            name: "String"
        }
    }
};
export const $host = {
    parameterPath: "$host",
    mapper: {
        serializedName: "$host",
        required: true,
        type: {
            name: "String"
        }
    },
    skipEncoding: true
};
export const apiVersion = {
    parameterPath: "apiVersion",
    mapper: {
        defaultValue: "2014-04-01",
        isConstant: true,
        serializedName: "api-version",
        type: {
            name: "String"
        }
    }
};
export const subscriptionId = {
    parameterPath: "subscriptionId",
    mapper: {
        serializedName: "subscriptionId",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const resourceGroupName = {
    parameterPath: "resourceGroupName",
    mapper: {
        serializedName: "resourceGroupName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const serverName = {
    parameterPath: "serverName",
    mapper: {
        serializedName: "serverName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const databaseName = {
    parameterPath: "databaseName",
    mapper: {
        serializedName: "databaseName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const contentType = {
    parameterPath: ["options", "contentType"],
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Content-Type",
        type: {
            name: "String"
        }
    }
};
export const parameters = {
    parameterPath: "parameters",
    mapper: DataMaskingPolicyMapper
};
export const dataMaskingPolicyName = {
    parameterPath: "dataMaskingPolicyName",
    mapper: {
        defaultValue: "Default",
        isConstant: true,
        serializedName: "dataMaskingPolicyName",
        type: {
            name: "String"
        }
    }
};
export const parameters1 = {
    parameterPath: "parameters",
    mapper: DataMaskingRuleMapper
};
export const dataMaskingRuleName = {
    parameterPath: "dataMaskingRuleName",
    mapper: {
        serializedName: "dataMaskingRuleName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters2 = {
    parameterPath: "parameters",
    mapper: GeoBackupPolicyMapper
};
export const geoBackupPolicyName = {
    parameterPath: "geoBackupPolicyName",
    mapper: {
        serializedName: "geoBackupPolicyName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const filter = {
    parameterPath: "filter",
    mapper: {
        serializedName: "$filter",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const skipToken = {
    parameterPath: ["options", "skipToken"],
    mapper: {
        serializedName: "$skipToken",
        type: {
            name: "String"
        }
    }
};
export const apiVersion1 = {
    parameterPath: "apiVersion",
    mapper: {
        defaultValue: "2021-05-01-preview",
        isConstant: true,
        serializedName: "api-version",
        type: {
            name: "String"
        }
    }
};
export const parameters3 = {
    parameterPath: "parameters",
    mapper: DatabaseMapper
};
export const parameters4 = {
    parameterPath: "parameters",
    mapper: DatabaseUpdateMapper
};
export const elasticPoolName = {
    parameterPath: "elasticPoolName",
    mapper: {
        serializedName: "elasticPoolName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const replicaType = {
    parameterPath: ["options", "replicaType"],
    mapper: {
        serializedName: "replicaType",
        type: {
            name: "String"
        }
    }
};
export const parameters5 = {
    parameterPath: "parameters",
    mapper: ResourceMoveDefinitionMapper
};
export const parameters6 = {
    parameterPath: "parameters",
    mapper: ImportExistingDatabaseDefinitionMapper
};
export const parameters7 = {
    parameterPath: "parameters",
    mapper: ExportDatabaseDefinitionMapper
};
export const nextLink = {
    parameterPath: "nextLink",
    mapper: {
        serializedName: "nextLink",
        required: true,
        type: {
            name: "String"
        }
    },
    skipEncoding: true
};
export const skip = {
    parameterPath: ["options", "skip"],
    mapper: {
        serializedName: "$skip",
        type: {
            name: "Number"
        }
    }
};
export const apiVersion2 = {
    parameterPath: "apiVersion",
    mapper: {
        defaultValue: "2020-11-01-preview",
        isConstant: true,
        serializedName: "api-version",
        type: {
            name: "String"
        }
    }
};
export const parameters8 = {
    parameterPath: "parameters",
    mapper: ElasticPoolMapper
};
export const parameters9 = {
    parameterPath: "parameters",
    mapper: ElasticPoolUpdateMapper
};
export const linkId = {
    parameterPath: "linkId",
    mapper: {
        serializedName: "linkId",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters10 = {
    parameterPath: "parameters",
    mapper: UnlinkParametersMapper
};
export const apiVersion3 = {
    parameterPath: "apiVersion",
    mapper: {
        defaultValue: "2021-02-01-preview",
        isConstant: true,
        serializedName: "api-version",
        type: {
            name: "String"
        }
    }
};
export const communicationLinkName = {
    parameterPath: "communicationLinkName",
    mapper: {
        serializedName: "communicationLinkName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters11 = {
    parameterPath: "parameters",
    mapper: ServerCommunicationLinkMapper
};
export const serviceObjectiveName = {
    parameterPath: "serviceObjectiveName",
    mapper: {
        serializedName: "serviceObjectiveName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const blobAuditingPolicyName = {
    parameterPath: "blobAuditingPolicyName",
    mapper: {
        defaultValue: "default",
        isConstant: true,
        serializedName: "blobAuditingPolicyName",
        type: {
            name: "String"
        }
    }
};
export const parameters12 = {
    parameterPath: "parameters",
    mapper: ExtendedDatabaseBlobAuditingPolicyMapper
};
export const parameters13 = {
    parameterPath: "parameters",
    mapper: ExtendedServerBlobAuditingPolicyMapper
};
export const parameters14 = {
    parameterPath: "parameters",
    mapper: ServerBlobAuditingPolicyMapper
};
export const parameters15 = {
    parameterPath: "parameters",
    mapper: DatabaseBlobAuditingPolicyMapper
};
export const expand = {
    parameterPath: ["options", "expand"],
    mapper: {
        serializedName: "$expand",
        type: {
            name: "String"
        }
    }
};
export const advisorName = {
    parameterPath: "advisorName",
    mapper: {
        serializedName: "advisorName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters16 = {
    parameterPath: "parameters",
    mapper: AdvisorMapper
};
export const parameters17 = {
    parameterPath: "parameters",
    mapper: DatabaseAutomaticTuningMapper
};
export const schema = {
    parameterPath: ["options", "schema"],
    mapper: {
        serializedName: "schema",
        type: {
            name: "Sequence",
            element: {
                type: {
                    name: "String"
                }
            }
        }
    },
    collectionFormat: "Multi"
};
export const table = {
    parameterPath: ["options", "table"],
    mapper: {
        serializedName: "table",
        type: {
            name: "Sequence",
            element: {
                type: {
                    name: "String"
                }
            }
        }
    },
    collectionFormat: "Multi"
};
export const column = {
    parameterPath: ["options", "column"],
    mapper: {
        serializedName: "column",
        type: {
            name: "Sequence",
            element: {
                type: {
                    name: "String"
                }
            }
        }
    },
    collectionFormat: "Multi"
};
export const orderBy = {
    parameterPath: ["options", "orderBy"],
    mapper: {
        serializedName: "orderBy",
        type: {
            name: "Sequence",
            element: {
                type: {
                    name: "String"
                }
            }
        }
    },
    collectionFormat: "Multi"
};
export const skiptoken = {
    parameterPath: ["options", "skiptoken"],
    mapper: {
        serializedName: "$skiptoken",
        type: {
            name: "String"
        }
    }
};
export const schemaName = {
    parameterPath: "schemaName",
    mapper: {
        serializedName: "schemaName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const tableName = {
    parameterPath: "tableName",
    mapper: {
        serializedName: "tableName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const filter1 = {
    parameterPath: ["options", "filter"],
    mapper: {
        serializedName: "$filter",
        type: {
            name: "String"
        }
    }
};
export const columnName = {
    parameterPath: "columnName",
    mapper: {
        serializedName: "columnName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const recommendedActionName = {
    parameterPath: "recommendedActionName",
    mapper: {
        serializedName: "recommendedActionName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters18 = {
    parameterPath: "parameters",
    mapper: RecommendedActionMapper
};
export const securityAlertPolicyName = {
    parameterPath: "securityAlertPolicyName",
    mapper: {
        serializedName: "securityAlertPolicyName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters19 = {
    parameterPath: "parameters",
    mapper: DatabaseSecurityAlertPolicyMapper
};
export const vulnerabilityAssessmentName = {
    parameterPath: "vulnerabilityAssessmentName",
    mapper: {
        serializedName: "vulnerabilityAssessmentName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const ruleId = {
    parameterPath: "ruleId",
    mapper: {
        serializedName: "ruleId",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const baselineName = {
    parameterPath: "baselineName",
    mapper: {
        serializedName: "baselineName",
        required: true,
        type: {
            name: "Enum",
            allowedValues: ["master", "default"]
        }
    }
};
export const parameters20 = {
    parameterPath: "parameters",
    mapper: DatabaseVulnerabilityAssessmentRuleBaselineMapper
};
export const parameters21 = {
    parameterPath: "parameters",
    mapper: DatabaseVulnerabilityAssessmentMapper
};
export const scanId = {
    parameterPath: "scanId",
    mapper: {
        serializedName: "scanId",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const dataWarehouseUserActivityName = {
    parameterPath: "dataWarehouseUserActivityName",
    mapper: {
        serializedName: "dataWarehouseUserActivityName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const locationName = {
    parameterPath: "locationName",
    mapper: {
        serializedName: "locationName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const deletedServerName = {
    parameterPath: "deletedServerName",
    mapper: {
        serializedName: "deletedServerName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const operationId = {
    parameterPath: "operationId",
    mapper: {
        serializedName: "operationId",
        required: true,
        type: {
            name: "Uuid"
        }
    }
};
export const encryptionProtectorName = {
    parameterPath: "encryptionProtectorName",
    mapper: {
        serializedName: "encryptionProtectorName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters22 = {
    parameterPath: "parameters",
    mapper: EncryptionProtectorMapper
};
export const failoverGroupName = {
    parameterPath: "failoverGroupName",
    mapper: {
        serializedName: "failoverGroupName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters23 = {
    parameterPath: "parameters",
    mapper: FailoverGroupMapper
};
export const parameters24 = {
    parameterPath: "parameters",
    mapper: FailoverGroupUpdateMapper
};
export const firewallRuleName = {
    parameterPath: "firewallRuleName",
    mapper: {
        serializedName: "firewallRuleName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters25 = {
    parameterPath: "parameters",
    mapper: FirewallRuleMapper
};
export const parameters26 = {
    parameterPath: "parameters",
    mapper: FirewallRuleListMapper
};
export const parameters27 = {
    parameterPath: "parameters",
    mapper: InstanceFailoverGroupMapper
};
export const instancePoolName = {
    parameterPath: "instancePoolName",
    mapper: {
        serializedName: "instancePoolName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters28 = {
    parameterPath: "parameters",
    mapper: InstancePoolMapper
};
export const parameters29 = {
    parameterPath: "parameters",
    mapper: InstancePoolUpdateMapper
};
export const jobAgentName = {
    parameterPath: "jobAgentName",
    mapper: {
        serializedName: "jobAgentName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters30 = {
    parameterPath: "parameters",
    mapper: JobAgentMapper
};
export const parameters31 = {
    parameterPath: "parameters",
    mapper: JobAgentUpdateMapper
};
export const credentialName = {
    parameterPath: "credentialName",
    mapper: {
        serializedName: "credentialName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters32 = {
    parameterPath: "parameters",
    mapper: JobCredentialMapper
};
export const createTimeMin = {
    parameterPath: ["options", "createTimeMin"],
    mapper: {
        serializedName: "createTimeMin",
        type: {
            name: "DateTime"
        }
    }
};
export const createTimeMax = {
    parameterPath: ["options", "createTimeMax"],
    mapper: {
        serializedName: "createTimeMax",
        type: {
            name: "DateTime"
        }
    }
};
export const endTimeMin = {
    parameterPath: ["options", "endTimeMin"],
    mapper: {
        serializedName: "endTimeMin",
        type: {
            name: "DateTime"
        }
    }
};
export const endTimeMax = {
    parameterPath: ["options", "endTimeMax"],
    mapper: {
        serializedName: "endTimeMax",
        type: {
            name: "DateTime"
        }
    }
};
export const isActive = {
    parameterPath: ["options", "isActive"],
    mapper: {
        serializedName: "isActive",
        type: {
            name: "Boolean"
        }
    }
};
export const top = {
    parameterPath: ["options", "top"],
    mapper: {
        serializedName: "$top",
        type: {
            name: "Number"
        }
    }
};
export const jobName = {
    parameterPath: "jobName",
    mapper: {
        serializedName: "jobName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const jobExecutionId = {
    parameterPath: "jobExecutionId",
    mapper: {
        serializedName: "jobExecutionId",
        required: true,
        type: {
            name: "Uuid"
        }
    }
};
export const parameters33 = {
    parameterPath: "parameters",
    mapper: JobMapper
};
export const stepName = {
    parameterPath: "stepName",
    mapper: {
        serializedName: "stepName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const jobVersion = {
    parameterPath: "jobVersion",
    mapper: {
        serializedName: "jobVersion",
        required: true,
        type: {
            name: "Number"
        }
    }
};
export const parameters34 = {
    parameterPath: "parameters",
    mapper: JobStepMapper
};
export const targetId = {
    parameterPath: "targetId",
    mapper: {
        serializedName: "targetId",
        required: true,
        type: {
            name: "Uuid"
        }
    }
};
export const targetGroupName = {
    parameterPath: "targetGroupName",
    mapper: {
        serializedName: "targetGroupName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters35 = {
    parameterPath: "parameters",
    mapper: JobTargetGroupMapper
};
export const include = {
    parameterPath: ["options", "include"],
    mapper: {
        serializedName: "include",
        type: {
            name: "String"
        }
    }
};
export const policyName = {
    parameterPath: "policyName",
    mapper: {
        serializedName: "policyName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters36 = {
    parameterPath: "parameters",
    mapper: LongTermRetentionPolicyMapper
};
export const maintenanceWindowOptionsName = {
    parameterPath: "maintenanceWindowOptionsName",
    mapper: {
        serializedName: "maintenanceWindowOptionsName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const maintenanceWindowName = {
    parameterPath: "maintenanceWindowName",
    mapper: {
        serializedName: "maintenanceWindowName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters37 = {
    parameterPath: "parameters",
    mapper: MaintenanceWindowsMapper
};
export const managedInstanceName = {
    parameterPath: "managedInstanceName",
    mapper: {
        serializedName: "managedInstanceName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const policyName1 = {
    parameterPath: "policyName",
    mapper: {
        serializedName: "policyName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters38 = {
    parameterPath: "parameters",
    mapper: ManagedBackupShortTermRetentionPolicyMapper
};
export const queryId = {
    parameterPath: "queryId",
    mapper: {
        serializedName: "queryId",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const startTime = {
    parameterPath: ["options", "startTime"],
    mapper: {
        serializedName: "startTime",
        type: {
            name: "String"
        }
    }
};
export const endTime = {
    parameterPath: ["options", "endTime"],
    mapper: {
        serializedName: "endTime",
        type: {
            name: "String"
        }
    }
};
export const interval = {
    parameterPath: ["options", "interval"],
    mapper: {
        serializedName: "interval",
        type: {
            name: "String"
        }
    }
};
export const restoreDetailsName = {
    parameterPath: "restoreDetailsName",
    mapper: {
        serializedName: "restoreDetailsName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters39 = {
    parameterPath: "parameters",
    mapper: ManagedDatabaseMapper
};
export const parameters40 = {
    parameterPath: "parameters",
    mapper: ManagedDatabaseUpdateMapper
};
export const parameters41 = {
    parameterPath: "parameters",
    mapper: CompleteDatabaseRestoreDefinitionMapper
};
export const parameters42 = {
    parameterPath: "parameters",
    mapper: ManagedDatabaseSecurityAlertPolicyMapper
};
export const sensitivityLabelSource = {
    parameterPath: "sensitivityLabelSource",
    mapper: {
        serializedName: "sensitivityLabelSource",
        required: true,
        type: {
            name: "Enum",
            allowedValues: ["current", "recommended"]
        }
    }
};
export const parameters43 = {
    parameterPath: "parameters",
    mapper: SensitivityLabelMapper
};
export const sensitivityLabelSource1 = {
    parameterPath: "sensitivityLabelSource",
    mapper: {
        defaultValue: "current",
        isConstant: true,
        serializedName: "sensitivityLabelSource",
        type: {
            name: "String"
        }
    }
};
export const sensitivityLabelSource2 = {
    parameterPath: "sensitivityLabelSource",
    mapper: {
        defaultValue: "recommended",
        isConstant: true,
        serializedName: "sensitivityLabelSource",
        type: {
            name: "String"
        }
    }
};
export const count = {
    parameterPath: ["options", "count"],
    mapper: {
        serializedName: "$count",
        type: {
            name: "Boolean"
        }
    }
};
export const parameters44 = {
    parameterPath: "parameters",
    mapper: SensitivityLabelUpdateListMapper
};
export const includeDisabledRecommendations = {
    parameterPath: ["options", "includeDisabledRecommendations"],
    mapper: {
        serializedName: "includeDisabledRecommendations",
        type: {
            name: "Boolean"
        }
    }
};
export const parameters45 = {
    parameterPath: "parameters",
    mapper: RecommendedSensitivityLabelUpdateListMapper
};
export const tdeName = {
    parameterPath: "tdeName",
    mapper: {
        serializedName: "tdeName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters46 = {
    parameterPath: "parameters",
    mapper: ManagedTransparentDataEncryptionMapper
};
export const administratorName = {
    parameterPath: "administratorName",
    mapper: {
        serializedName: "administratorName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters47 = {
    parameterPath: "parameters",
    mapper: ManagedInstanceAdministratorMapper
};
export const authenticationName = {
    parameterPath: "authenticationName",
    mapper: {
        serializedName: "authenticationName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters48 = {
    parameterPath: "parameters",
    mapper: ManagedInstanceAzureADOnlyAuthenticationMapper
};
export const parameters49 = {
    parameterPath: "parameters",
    mapper: ManagedInstanceEncryptionProtectorMapper
};
export const keyName = {
    parameterPath: "keyName",
    mapper: {
        serializedName: "keyName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters50 = {
    parameterPath: "parameters",
    mapper: ManagedInstanceKeyMapper
};
export const policyName2 = {
    parameterPath: "policyName",
    mapper: {
        serializedName: "policyName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters51 = {
    parameterPath: "parameters",
    mapper: ManagedInstanceLongTermRetentionPolicyMapper
};
export const privateEndpointConnectionName = {
    parameterPath: "privateEndpointConnectionName",
    mapper: {
        serializedName: "privateEndpointConnectionName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters52 = {
    parameterPath: "parameters",
    mapper: ManagedInstancePrivateEndpointConnectionMapper
};
export const groupName = {
    parameterPath: "groupName",
    mapper: {
        serializedName: "groupName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters53 = {
    parameterPath: "parameters",
    mapper: TdeCertificateMapper
};
export const parameters54 = {
    parameterPath: "parameters",
    mapper: ManagedInstanceVulnerabilityAssessmentMapper
};
export const restorableDroppedDatabaseId = {
    parameterPath: "restorableDroppedDatabaseId",
    mapper: {
        serializedName: "restorableDroppedDatabaseId",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters55 = {
    parameterPath: "parameters",
    mapper: ManagedServerSecurityAlertPolicyMapper
};
export const parameters56 = {
    parameterPath: "parameters",
    mapper: PrivateEndpointConnectionMapper
};
export const recoverableDatabaseName = {
    parameterPath: "recoverableDatabaseName",
    mapper: {
        serializedName: "recoverableDatabaseName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters57 = {
    parameterPath: "parameters",
    mapper: CreateDatabaseRestorePointDefinitionMapper
};
export const restorePointName = {
    parameterPath: "restorePointName",
    mapper: {
        serializedName: "restorePointName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters58 = {
    parameterPath: "parameters",
    mapper: ServerAutomaticTuningMapper
};
export const parameters59 = {
    parameterPath: "parameters",
    mapper: ServerAzureADAdministratorMapper
};
export const parameters60 = {
    parameterPath: "parameters",
    mapper: ServerAzureADOnlyAuthenticationMapper
};
export const devOpsAuditingSettingsName = {
    parameterPath: "devOpsAuditingSettingsName",
    mapper: {
        serializedName: "devOpsAuditingSettingsName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters61 = {
    parameterPath: "parameters",
    mapper: ServerDevOpsAuditingSettingsMapper
};
export const dnsAliasName = {
    parameterPath: "dnsAliasName",
    mapper: {
        serializedName: "dnsAliasName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters62 = {
    parameterPath: "parameters",
    mapper: ServerDnsAliasAcquisitionMapper
};
export const parameters63 = {
    parameterPath: "parameters",
    mapper: ServerKeyMapper
};
export const parameters64 = {
    parameterPath: "parameters",
    mapper: ServerSecurityAlertPolicyMapper
};
export const serverTrustGroupName = {
    parameterPath: "serverTrustGroupName",
    mapper: {
        serializedName: "serverTrustGroupName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters65 = {
    parameterPath: "parameters",
    mapper: ServerTrustGroupMapper
};
export const parameters66 = {
    parameterPath: "parameters",
    mapper: ServerVulnerabilityAssessmentMapper
};
export const parameters67 = {
    parameterPath: "parameters",
    mapper: SqlAgentConfigurationMapper
};
export const usageName = {
    parameterPath: "usageName",
    mapper: {
        serializedName: "usageName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const syncAgentName = {
    parameterPath: "syncAgentName",
    mapper: {
        serializedName: "syncAgentName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters68 = {
    parameterPath: "parameters",
    mapper: SyncAgentMapper
};
export const syncGroupName = {
    parameterPath: "syncGroupName",
    mapper: {
        serializedName: "syncGroupName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const startTime1 = {
    parameterPath: "startTime",
    mapper: {
        serializedName: "startTime",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const endTime1 = {
    parameterPath: "endTime",
    mapper: {
        serializedName: "endTime",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const typeParam = {
    parameterPath: "typeParam",
    mapper: {
        serializedName: "type",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const continuationToken = {
    parameterPath: ["options", "continuationToken"],
    mapper: {
        serializedName: "continuationToken",
        type: {
            name: "String"
        }
    }
};
export const parameters69 = {
    parameterPath: "parameters",
    mapper: SyncGroupMapper
};
export const syncMemberName = {
    parameterPath: "syncMemberName",
    mapper: {
        serializedName: "syncMemberName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters70 = {
    parameterPath: "parameters",
    mapper: SyncMemberMapper
};
export const timeZoneId = {
    parameterPath: "timeZoneId",
    mapper: {
        serializedName: "timeZoneId",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const virtualClusterName = {
    parameterPath: "virtualClusterName",
    mapper: {
        serializedName: "virtualClusterName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters71 = {
    parameterPath: "parameters",
    mapper: VirtualClusterUpdateMapper
};
export const virtualNetworkRuleName = {
    parameterPath: "virtualNetworkRuleName",
    mapper: {
        serializedName: "virtualNetworkRuleName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters72 = {
    parameterPath: "parameters",
    mapper: VirtualNetworkRuleMapper
};
export const workloadGroupName = {
    parameterPath: "workloadGroupName",
    mapper: {
        serializedName: "workloadGroupName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const workloadClassifierName = {
    parameterPath: "workloadClassifierName",
    mapper: {
        serializedName: "workloadClassifierName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters73 = {
    parameterPath: "parameters",
    mapper: WorkloadClassifierMapper
};
export const parameters74 = {
    parameterPath: "parameters",
    mapper: WorkloadGroupMapper
};
export const parameters75 = {
    parameterPath: "parameters",
    mapper: LogicalDatabaseTransparentDataEncryptionMapper
};
export const policyName3 = {
    parameterPath: "policyName",
    mapper: {
        serializedName: "policyName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters76 = {
    parameterPath: "parameters",
    mapper: BackupShortTermRetentionPolicyMapper
};
export const extensionName = {
    parameterPath: "extensionName",
    mapper: {
        serializedName: "extensionName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters77 = {
    parameterPath: "parameters",
    mapper: DatabaseExtensionsMapper
};
export const ledgerDigestUploads = {
    parameterPath: "ledgerDigestUploads",
    mapper: {
        serializedName: "ledgerDigestUploads",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters78 = {
    parameterPath: "parameters",
    mapper: LedgerDigestUploadsMapper
};
export const outboundRuleFqdn = {
    parameterPath: "outboundRuleFqdn",
    mapper: {
        serializedName: "outboundRuleFqdn",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters79 = {
    parameterPath: "parameters",
    mapper: OutboundFirewallRuleMapper
};
export const parameters80 = {
    parameterPath: "parameters",
    mapper: ServerMapper
};
export const parameters81 = {
    parameterPath: "parameters",
    mapper: ServerUpdateMapper
};
export const parameters82 = {
    parameterPath: "parameters",
    mapper: ImportNewDatabaseDefinitionMapper
};
export const parameters83 = {
    parameterPath: "parameters",
    mapper: CheckNameAvailabilityRequestMapper
};
export const expandChildren = {
    parameterPath: ["options", "expandChildren"],
    mapper: {
        serializedName: "expandChildren",
        type: {
            name: "Boolean"
        }
    }
};
export const parameters84 = {
    parameterPath: "parameters",
    mapper: CopyLongTermRetentionBackupParametersMapper
};
export const longTermRetentionServerName = {
    parameterPath: "longTermRetentionServerName",
    mapper: {
        serializedName: "longTermRetentionServerName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const longTermRetentionDatabaseName = {
    parameterPath: "longTermRetentionDatabaseName",
    mapper: {
        serializedName: "longTermRetentionDatabaseName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const backupName = {
    parameterPath: "backupName",
    mapper: {
        serializedName: "backupName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters85 = {
    parameterPath: "parameters",
    mapper: UpdateLongTermRetentionBackupParametersMapper
};
export const onlyLatestPerDatabase = {
    parameterPath: ["options", "onlyLatestPerDatabase"],
    mapper: {
        serializedName: "onlyLatestPerDatabase",
        type: {
            name: "Boolean"
        }
    }
};
export const databaseState = {
    parameterPath: ["options", "databaseState"],
    mapper: {
        serializedName: "databaseState",
        type: {
            name: "String"
        }
    }
};
export const parameters86 = {
    parameterPath: "parameters",
    mapper: ManagedInstanceMapper
};
export const parameters87 = {
    parameterPath: "parameters",
    mapper: ManagedInstanceUpdateMapper
};
export const numberOfQueries = {
    parameterPath: ["options", "numberOfQueries"],
    mapper: {
        serializedName: "numberOfQueries",
        type: {
            name: "Number"
        }
    }
};
export const databases = {
    parameterPath: ["options", "databases"],
    mapper: {
        serializedName: "databases",
        type: {
            name: "String"
        }
    }
};
export const aggregationFunction = {
    parameterPath: ["options", "aggregationFunction"],
    mapper: {
        serializedName: "aggregationFunction",
        type: {
            name: "String"
        }
    }
};
export const observationMetric = {
    parameterPath: ["options", "observationMetric"],
    mapper: {
        serializedName: "observationMetric",
        type: {
            name: "String"
        }
    }
};
export const connectionPolicyName = {
    parameterPath: "connectionPolicyName",
    mapper: {
        serializedName: "connectionPolicyName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters88 = {
    parameterPath: "parameters",
    mapper: ServerConnectionPolicyMapper
};
