// @fluentui/react-field/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Field: function() {
        return _Field.Field;
    },
    fieldClassNames: function() {
        return _Field.fieldClassNames;
    },
    renderField_unstable: function() {
        return _Field.renderField_unstable;
    },
    useFieldStyles_unstable: function() {
        return _Field.useFieldStyles_unstable;
    },
    useField_unstable: function() {
        return _Field.useField_unstable;
    },
    FieldContextProvider: function() {
        return _index.FieldContextProvider;
    },
    useFieldContext_unstable: function() {
        return _index.useFieldContext_unstable;
    },
    useFieldContextValues_unstable: function() {
        return _index.useFieldContextValues_unstable;
    },
    useFieldControlProps_unstable: function() {
        return _index.useFieldControlProps_unstable;
    }
});
const _Field = require("./Field");
const _index = require("./contexts/index");
