// @fluentui/react-accordion/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Accordion: function() {
        return _Accordion.Accordion;
    },
    accordionClassNames: function() {
        return _Accordion.accordionClassNames;
    },
    renderAccordion_unstable: function() {
        return _Accordion.renderAccordion_unstable;
    },
    useAccordionContextValues_unstable: function() {
        return _Accordion.useAccordionContextValues_unstable;
    },
    useAccordionStyles_unstable: function() {
        return _Accordion.useAccordionStyles_unstable;
    },
    useAccordion_unstable: function() {
        return _Accordion.useAccordion_unstable;
    },
    AccordionItem: function() {
        return _AccordionItem.AccordionItem;
    },
    accordionItemClassNames: function() {
        return _AccordionItem.accordionItemClassNames;
    },
    renderAccordionItem_unstable: function() {
        return _AccordionItem.renderAccordionItem_unstable;
    },
    useAccordionItemContextValues_unstable: function() {
        return _AccordionItem.useAccordionItemContextValues_unstable;
    },
    useAccordionItemStyles_unstable: function() {
        return _AccordionItem.useAccordionItemStyles_unstable;
    },
    useAccordionItem_unstable: function() {
        return _AccordionItem.useAccordionItem_unstable;
    },
    AccordionHeader: function() {
        return _AccordionHeader.AccordionHeader;
    },
    accordionHeaderClassNames: function() {
        return _AccordionHeader.accordionHeaderClassNames;
    },
    renderAccordionHeader_unstable: function() {
        return _AccordionHeader.renderAccordionHeader_unstable;
    },
    useAccordionHeaderContextValues_unstable: function() {
        return _AccordionHeader.useAccordionHeaderContextValues_unstable;
    },
    useAccordionHeaderStyles_unstable: function() {
        return _AccordionHeader.useAccordionHeaderStyles_unstable;
    },
    useAccordionHeader_unstable: function() {
        return _AccordionHeader.useAccordionHeader_unstable;
    },
    AccordionPanel: function() {
        return _AccordionPanel.AccordionPanel;
    },
    accordionPanelClassNames: function() {
        return _AccordionPanel.accordionPanelClassNames;
    },
    renderAccordionPanel_unstable: function() {
        return _AccordionPanel.renderAccordionPanel_unstable;
    },
    useAccordionPanelStyles_unstable: function() {
        return _AccordionPanel.useAccordionPanelStyles_unstable;
    },
    useAccordionPanel_unstable: function() {
        return _AccordionPanel.useAccordionPanel_unstable;
    },
    AccordionProvider: function() {
        return _accordion.AccordionProvider;
    },
    useAccordionContext_unstable: function() {
        return _accordion.useAccordionContext_unstable;
    },
    AccordionItemProvider: function() {
        return _accordionItem.AccordionItemProvider;
    },
    useAccordionItemContext_unstable: function() {
        return _accordionItem.useAccordionItemContext_unstable;
    },
    AccordionHeaderProvider: function() {
        return _accordionHeader.AccordionHeaderProvider;
    },
    useAccordionHeaderContext_unstable: function() {
        return _accordionHeader.useAccordionHeaderContext_unstable;
    }
});
const _Accordion = require("./Accordion");
const _AccordionItem = require("./AccordionItem");
const _AccordionHeader = require("./AccordionHeader");
const _AccordionPanel = require("./AccordionPanel");
const _accordion = require("./contexts/accordion");
const _accordionItem = require("./contexts/accordionItem");
const _accordionHeader = require("./contexts/accordionHeader");
