// @fluentui/react-card/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Card: function() {
        return _Card.Card;
    },
    cardClassNames: function() {
        return _Card.cardClassNames;
    },
    cardCSSVars: function() {
        return _Card.cardCSSVars;
    },
    renderCard_unstable: function() {
        return _Card.renderCard_unstable;
    },
    useCardStyles_unstable: function() {
        return _Card.useCardStyles_unstable;
    },
    useCard_unstable: function() {
        return _Card.useCard_unstable;
    },
    CardFooter: function() {
        return _CardFooter.CardFooter;
    },
    cardFooterClassNames: function() {
        return _CardFooter.cardFooterClassNames;
    },
    renderCardFooter_unstable: function() {
        return _CardFooter.renderCardFooter_unstable;
    },
    useCardFooterStyles_unstable: function() {
        return _CardFooter.useCardFooterStyles_unstable;
    },
    useCardFooter_unstable: function() {
        return _CardFooter.useCardFooter_unstable;
    },
    CardHeader: function() {
        return _CardHeader.CardHeader;
    },
    cardHeaderClassNames: function() {
        return _CardHeader.cardHeaderClassNames;
    },
    cardHeaderCSSVars: function() {
        return _CardHeader.cardHeaderCSSVars;
    },
    renderCardHeader_unstable: function() {
        return _CardHeader.renderCardHeader_unstable;
    },
    useCardHeaderStyles_unstable: function() {
        return _CardHeader.useCardHeaderStyles_unstable;
    },
    useCardHeader_unstable: function() {
        return _CardHeader.useCardHeader_unstable;
    },
    CardPreview: function() {
        return _CardPreview.CardPreview;
    },
    cardPreviewClassNames: function() {
        return _CardPreview.cardPreviewClassNames;
    },
    renderCardPreview_unstable: function() {
        return _CardPreview.renderCardPreview_unstable;
    },
    useCardPreviewStyles_unstable: function() {
        return _CardPreview.useCardPreviewStyles_unstable;
    },
    useCardPreview_unstable: function() {
        return _CardPreview.useCardPreview_unstable;
    },
    CardProvider: function() {
        return _Card.CardProvider;
    },
    useCardContext_unstable: function() {
        return _Card.useCardContext_unstable;
    }
});
const _Card = require("./Card");
const _CardFooter = require("./CardFooter");
const _CardHeader = require("./CardHeader");
const _CardPreview = require("./CardPreview");
