// @fluentui/react-infobutton/lib-commonjs/index.js
/* eslint-disable deprecation/deprecation */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    InfoButton: function() {
        return _InfoButton.InfoButton;
    },
    infoButtonClassNames: function() {
        return _InfoButton.infoButtonClassNames;
    },
    renderInfoButton_unstable: function() {
        return _InfoButton.renderInfoButton_unstable;
    },
    useInfoButtonStyles_unstable: function() {
        return _InfoButton.useInfoButtonStyles_unstable;
    },
    useInfoButton_unstable: function() {
        return _InfoButton.useInfoButton_unstable;
    },
    InfoLabel: function() {
        return _InfoLabel.InfoLabel;
    },
    infoLabelClassNames: function() {
        return _InfoLabel.infoLabelClassNames;
    },
    renderInfoLabel_unstable: function() {
        return _InfoLabel.renderInfoLabel_unstable;
    },
    useInfoLabelStyles_unstable: function() {
        return _InfoLabel.useInfoLabelStyles_unstable;
    },
    useInfoLabel_unstable: function() {
        return _InfoLabel.useInfoLabel_unstable;
    }
});
const _InfoButton = require("./InfoButton");
const _InfoLabel = require("./InfoLabel");
