// @fluentui/react-aria/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    useARIAButtonShorthand: function() {
        return _index.useARIAButtonShorthand;
    },
    useARIAButtonProps: function() {
        return _index.useARIAButtonProps;
    },
    useActiveDescendant: function() {
        return _activedescendant.useActiveDescendant;
    },
    ACTIVEDESCENDANT_FOCUSVISIBLE_ATTRIBUTE: function() {
        return _activedescendant.ACTIVEDESCENDANT_FOCUSVISIBLE_ATTRIBUTE;
    },
    AriaLiveAnnouncer: function() {
        return _index1.AriaLiveAnnouncer;
    },
    renderAriaLiveAnnouncer_unstable: function() {
        return _index1.renderAriaLiveAnnouncer_unstable;
    },
    useAriaLiveAnnouncer_unstable: function() {
        return _index1.useAriaLiveAnnouncer_unstable;
    },
    useAriaLiveAnnouncerContextValues_unstable: function() {
        return _index1.useAriaLiveAnnouncerContextValues_unstable;
    }
});
const _index = require("./button/index");
const _activedescendant = require("./activedescendant");
const _index1 = require("./AriaLiveAnnouncer/index");
