// @fluentui/react-menu/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    MenuProvider: function() {
        return _menuContext.MenuProvider;
    },
    useMenuContext_unstable: function() {
        return _menuContext.useMenuContext_unstable;
    },
    MenuTriggerContextProvider: function() {
        return _menuTriggerContext.MenuTriggerContextProvider;
    },
    useMenuTriggerContext_unstable: function() {
        return _menuTriggerContext.useMenuTriggerContext_unstable;
    },
    MenuGroupContextProvider: function() {
        return _menuGroupContext.MenuGroupContextProvider;
    },
    useMenuGroupContext_unstable: function() {
        return _menuGroupContext.useMenuGroupContext_unstable;
    },
    MenuListProvider: function() {
        return _menuListContext.MenuListProvider;
    },
    useMenuListContext_unstable: function() {
        return _menuListContext.useMenuListContext_unstable;
    },
    Menu: function() {
        return _Menu.Menu;
    },
    renderMenu_unstable: function() {
        return _Menu.renderMenu_unstable;
    },
    useMenuContextValues_unstable: function() {
        return _Menu.useMenuContextValues_unstable;
    },
    useMenu_unstable: function() {
        return _Menu.useMenu_unstable;
    },
    MenuDivider: function() {
        return _MenuDivider.MenuDivider;
    },
    menuDividerClassNames: function() {
        return _MenuDivider.menuDividerClassNames;
    },
    renderMenuDivider_unstable: function() {
        return _MenuDivider.renderMenuDivider_unstable;
    },
    useMenuDividerStyles_unstable: function() {
        return _MenuDivider.useMenuDividerStyles_unstable;
    },
    useMenuDivider_unstable: function() {
        return _MenuDivider.useMenuDivider_unstable;
    },
    MenuGroup: function() {
        return _MenuGroup.MenuGroup;
    },
    menuGroupClassNames: function() {
        return _MenuGroup.menuGroupClassNames;
    },
    renderMenuGroup_unstable: function() {
        return _MenuGroup.renderMenuGroup_unstable;
    },
    useMenuGroupContextValues_unstable: function() {
        return _MenuGroup.useMenuGroupContextValues_unstable;
    },
    useMenuGroupStyles_unstable: function() {
        return _MenuGroup.useMenuGroupStyles_unstable;
    },
    useMenuGroup_unstable: function() {
        return _MenuGroup.useMenuGroup_unstable;
    },
    MenuGroupHeader: function() {
        return _MenuGroupHeader.MenuGroupHeader;
    },
    menuGroupHeaderClassNames: function() {
        return _MenuGroupHeader.menuGroupHeaderClassNames;
    },
    renderMenuGroupHeader_unstable: function() {
        return _MenuGroupHeader.renderMenuGroupHeader_unstable;
    },
    useMenuGroupHeaderStyles_unstable: function() {
        return _MenuGroupHeader.useMenuGroupHeaderStyles_unstable;
    },
    useMenuGroupHeader_unstable: function() {
        return _MenuGroupHeader.useMenuGroupHeader_unstable;
    },
    MenuItem: function() {
        return _MenuItem.MenuItem;
    },
    menuItemClassNames: function() {
        return _MenuItem.menuItemClassNames;
    },
    renderMenuItem_unstable: function() {
        return _MenuItem.renderMenuItem_unstable;
    },
    useMenuItemStyles_unstable: function() {
        return _MenuItem.useMenuItemStyles_unstable;
    },
    useMenuItem_unstable: function() {
        return _MenuItem.useMenuItem_unstable;
    },
    MenuItemCheckbox: function() {
        return _MenuItemCheckbox.MenuItemCheckbox;
    },
    menuItemCheckboxClassNames: function() {
        return _MenuItemCheckbox.menuItemCheckboxClassNames;
    },
    renderMenuItemCheckbox_unstable: function() {
        return _MenuItemCheckbox.renderMenuItemCheckbox_unstable;
    },
    useMenuItemCheckboxStyles_unstable: function() {
        return _MenuItemCheckbox.useMenuItemCheckboxStyles_unstable;
    },
    useMenuItemCheckbox_unstable: function() {
        return _MenuItemCheckbox.useMenuItemCheckbox_unstable;
    },
    MenuItemRadio: function() {
        return _MenuItemRadio.MenuItemRadio;
    },
    menuItemRadioClassNames: function() {
        return _MenuItemRadio.menuItemRadioClassNames;
    },
    renderMenuItemRadio_unstable: function() {
        return _MenuItemRadio.renderMenuItemRadio_unstable;
    },
    useMenuItemRadioStyles_unstable: function() {
        return _MenuItemRadio.useMenuItemRadioStyles_unstable;
    },
    useMenuItemRadio_unstable: function() {
        return _MenuItemRadio.useMenuItemRadio_unstable;
    },
    MenuList: function() {
        return _MenuList.MenuList;
    },
    menuListClassNames: function() {
        return _MenuList.menuListClassNames;
    },
    renderMenuList_unstable: function() {
        return _MenuList.renderMenuList_unstable;
    },
    useMenuListContextValues_unstable: function() {
        return _MenuList.useMenuListContextValues_unstable;
    },
    useMenuListStyles_unstable: function() {
        return _MenuList.useMenuListStyles_unstable;
    },
    useMenuList_unstable: function() {
        return _MenuList.useMenuList_unstable;
    },
    MenuPopover: function() {
        return _MenuPopover.MenuPopover;
    },
    menuPopoverClassNames: function() {
        return _MenuPopover.menuPopoverClassNames;
    },
    renderMenuPopover_unstable: function() {
        return _MenuPopover.renderMenuPopover_unstable;
    },
    useMenuPopoverStyles_unstable: function() {
        return _MenuPopover.useMenuPopoverStyles_unstable;
    },
    useMenuPopover_unstable: function() {
        return _MenuPopover.useMenuPopover_unstable;
    },
    MenuSplitGroup: function() {
        return _MenuSplitGroup.MenuSplitGroup;
    },
    menuSplitGroupClassNames: function() {
        return _MenuSplitGroup.menuSplitGroupClassNames;
    },
    renderMenuSplitGroup_unstable: function() {
        return _MenuSplitGroup.renderMenuSplitGroup_unstable;
    },
    useMenuSplitGroupStyles_unstable: function() {
        return _MenuSplitGroup.useMenuSplitGroupStyles_unstable;
    },
    useMenuSplitGroup_unstable: function() {
        return _MenuSplitGroup.useMenuSplitGroup_unstable;
    },
    MenuTrigger: function() {
        return _MenuTrigger.MenuTrigger;
    },
    renderMenuTrigger_unstable: function() {
        return _MenuTrigger.renderMenuTrigger_unstable;
    },
    useMenuTrigger_unstable: function() {
        return _MenuTrigger.useMenuTrigger_unstable;
    },
    useCheckmarkStyles_unstable: function() {
        return _index.useCheckmarkStyles_unstable;
    },
    MenuItemLink: function() {
        return _MenuItemLink.MenuItemLink;
    },
    menuItemLinkClassNames: function() {
        return _MenuItemLink.menuItemLinkClassNames;
    },
    renderMenuItemLink_unstable: function() {
        return _MenuItemLink.renderMenuItemLink_unstable;
    },
    useMenuItemLinkStyles_unstable: function() {
        return _MenuItemLink.useMenuItemLinkStyles_unstable;
    },
    useMenuItemLink_unstable: function() {
        return _MenuItemLink.useMenuItemLink_unstable;
    },
    MENU_ENTER_EVENT: function() {
        return _utils.MENU_ENTER_EVENT;
    },
    dispatchMenuEnterEvent: function() {
        return _utils.dispatchMenuEnterEvent;
    },
    useOnMenuMouseEnter: function() {
        return _utils.useOnMenuMouseEnter;
    }
});
const _menuContext = require("./contexts/menuContext");
const _menuTriggerContext = require("./contexts/menuTriggerContext");
const _menuGroupContext = require("./contexts/menuGroupContext");
const _menuListContext = require("./contexts/menuListContext");
const _Menu = require("./Menu");
const _MenuDivider = require("./MenuDivider");
const _MenuGroup = require("./MenuGroup");
const _MenuGroupHeader = require("./MenuGroupHeader");
const _MenuItem = require("./MenuItem");
const _MenuItemCheckbox = require("./MenuItemCheckbox");
const _MenuItemRadio = require("./MenuItemRadio");
const _MenuList = require("./MenuList");
const _MenuPopover = require("./MenuPopover");
const _MenuSplitGroup = require("./MenuSplitGroup");
const _MenuTrigger = require("./MenuTrigger");
const _index = require("./selectable/index");
const _MenuItemLink = require("./MenuItemLink");
const _utils = require("./utils");
