// @azure/arm-sql/dist-esm/src/operations/managedDatabaseQueries.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { __asyncDelegator, __asyncGenerator, __asyncValues, __await } from "tslib";
import { setContinuationToken } from "../pagingHelper";
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers";
import * as Parameters from "../models/parameters";
/// <reference lib="esnext.asynciterable" />
/** Class containing ManagedDatabaseQueries operations. */
export class ManagedDatabaseQueriesImpl {
    /**
     * Initialize a new instance of the class ManagedDatabaseQueries class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Get query execution statistics by query id.
     * @param resourceGroupName The name of the resource group that contains the resource. You can obtain
     *                          this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param databaseName The name of the database.
     * @param queryId
     * @param options The options parameters.
     */
    listByQuery(resourceGroupName, managedInstanceName, databaseName, queryId, options) {
        const iter = this.listByQueryPagingAll(resourceGroupName, managedInstanceName, databaseName, queryId, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings === null || settings === void 0 ? void 0 : settings.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listByQueryPagingPage(resourceGroupName, managedInstanceName, databaseName, queryId, options, settings);
            }
        };
    }
    listByQueryPagingPage(resourceGroupName, managedInstanceName, databaseName, queryId, options, settings) {
        return __asyncGenerator(this, arguments, function* listByQueryPagingPage_1() {
            let result;
            let continuationToken = settings === null || settings === void 0 ? void 0 : settings.continuationToken;
            if (!continuationToken) {
                result = yield __await(this._listByQuery(resourceGroupName, managedInstanceName, databaseName, queryId, options));
                let page = result.value || [];
                continuationToken = result.nextLink;
                setContinuationToken(page, continuationToken);
                yield yield __await(page);
            }
            while (continuationToken) {
                result = yield __await(this._listByQueryNext(resourceGroupName, managedInstanceName, databaseName, queryId, continuationToken, options));
                continuationToken = result.nextLink;
                let page = result.value || [];
                setContinuationToken(page, continuationToken);
                yield yield __await(page);
            }
        });
    }
    listByQueryPagingAll(resourceGroupName, managedInstanceName, databaseName, queryId, options) {
        return __asyncGenerator(this, arguments, function* listByQueryPagingAll_1() {
            var e_1, _a;
            try {
                for (var _b = __asyncValues(this.listByQueryPagingPage(resourceGroupName, managedInstanceName, databaseName, queryId, options)), _c; _c = yield __await(_b.next()), !_c.done;) {
                    const page = _c.value;
                    yield __await(yield* __asyncDelegator(__asyncValues(page)));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) yield __await(_a.call(_b));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    /**
     * Get query by query id.
     * @param resourceGroupName The name of the resource group that contains the resource. You can obtain
     *                          this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param databaseName The name of the database.
     * @param queryId
     * @param options The options parameters.
     */
    get(resourceGroupName, managedInstanceName, databaseName, queryId, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            managedInstanceName,
            databaseName,
            queryId,
            options
        }, getOperationSpec);
    }
    /**
     * Get query execution statistics by query id.
     * @param resourceGroupName The name of the resource group that contains the resource. You can obtain
     *                          this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param databaseName The name of the database.
     * @param queryId
     * @param options The options parameters.
     */
    _listByQuery(resourceGroupName, managedInstanceName, databaseName, queryId, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            managedInstanceName,
            databaseName,
            queryId,
            options
        }, listByQueryOperationSpec);
    }
    /**
     * ListByQueryNext
     * @param resourceGroupName The name of the resource group that contains the resource. You can obtain
     *                          this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param databaseName The name of the database.
     * @param queryId
     * @param nextLink The nextLink from the previous successful call to the ListByQuery method.
     * @param options The options parameters.
     */
    _listByQueryNext(resourceGroupName, managedInstanceName, databaseName, queryId, nextLink, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            managedInstanceName,
            databaseName,
            queryId,
            nextLink,
            options
        }, listByQueryNextOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/databases/{databaseName}/queries/{queryId}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ManagedInstanceQuery
        },
        default: {}
    },
    queryParameters: [Parameters.apiVersion2],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.databaseName,
        Parameters.managedInstanceName,
        Parameters.queryId
    ],
    headerParameters: [Parameters.accept],
    serializer
};
const listByQueryOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/databases/{databaseName}/queries/{queryId}/statistics",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ManagedInstanceQueryStatistics
        },
        default: {}
    },
    queryParameters: [
        Parameters.apiVersion2,
        Parameters.startTime,
        Parameters.endTime,
        Parameters.interval
    ],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.databaseName,
        Parameters.managedInstanceName,
        Parameters.queryId
    ],
    headerParameters: [Parameters.accept],
    serializer
};
const listByQueryNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ManagedInstanceQueryStatistics
        },
        default: {}
    },
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.databaseName,
        Parameters.nextLink,
        Parameters.managedInstanceName,
        Parameters.queryId
    ],
    headerParameters: [Parameters.accept],
    serializer
};
