// @fluentui/react-drawer/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Drawer: function() {
        return _Drawer.Drawer;
    },
    renderDrawer_unstable: function() {
        return _Drawer.renderDrawer_unstable;
    },
    useDrawerStyles_unstable: function() {
        return _Drawer.useDrawerStyles_unstable;
    },
    useDrawer_unstable: function() {
        return _Drawer.useDrawer_unstable;
    },
    OverlayDrawer: function() {
        return _OverlayDrawer.OverlayDrawer;
    },
    overlayDrawerClassNames: function() {
        return _OverlayDrawer.overlayDrawerClassNames;
    },
    renderOverlayDrawer_unstable: function() {
        return _OverlayDrawer.renderOverlayDrawer_unstable;
    },
    useOverlayDrawerStyles_unstable: function() {
        return _OverlayDrawer.useOverlayDrawerStyles_unstable;
    },
    useOverlayDrawer_unstable: function() {
        return _OverlayDrawer.useOverlayDrawer_unstable;
    },
    InlineDrawer: function() {
        return _InlineDrawer.InlineDrawer;
    },
    inlineDrawerClassNames: function() {
        return _InlineDrawer.inlineDrawerClassNames;
    },
    renderInlineDrawer_unstable: function() {
        return _InlineDrawer.renderInlineDrawer_unstable;
    },
    useInlineDrawerStyles_unstable: function() {
        return _InlineDrawer.useInlineDrawerStyles_unstable;
    },
    useInlineDrawer_unstable: function() {
        return _InlineDrawer.useInlineDrawer_unstable;
    },
    DrawerBody: function() {
        return _DrawerBody.DrawerBody;
    },
    drawerBodyClassNames: function() {
        return _DrawerBody.drawerBodyClassNames;
    },
    renderDrawerBody_unstable: function() {
        return _DrawerBody.renderDrawerBody_unstable;
    },
    useDrawerBodyStyles_unstable: function() {
        return _DrawerBody.useDrawerBodyStyles_unstable;
    },
    useDrawerBody_unstable: function() {
        return _DrawerBody.useDrawerBody_unstable;
    },
    DrawerHeader: function() {
        return _DrawerHeader.DrawerHeader;
    },
    drawerHeaderClassNames: function() {
        return _DrawerHeader.drawerHeaderClassNames;
    },
    renderDrawerHeader_unstable: function() {
        return _DrawerHeader.renderDrawerHeader_unstable;
    },
    useDrawerHeaderStyles_unstable: function() {
        return _DrawerHeader.useDrawerHeaderStyles_unstable;
    },
    useDrawerHeader_unstable: function() {
        return _DrawerHeader.useDrawerHeader_unstable;
    },
    DrawerHeaderTitle: function() {
        return _DrawerHeaderTitle.DrawerHeaderTitle;
    },
    drawerHeaderTitleClassNames: function() {
        return _DrawerHeaderTitle.drawerHeaderTitleClassNames;
    },
    renderDrawerHeaderTitle_unstable: function() {
        return _DrawerHeaderTitle.renderDrawerHeaderTitle_unstable;
    },
    useDrawerHeaderTitleStyles_unstable: function() {
        return _DrawerHeaderTitle.useDrawerHeaderTitleStyles_unstable;
    },
    useDrawerHeaderTitle_unstable: function() {
        return _DrawerHeaderTitle.useDrawerHeaderTitle_unstable;
    },
    DrawerHeaderNavigation: function() {
        return _DrawerHeaderNavigation.DrawerHeaderNavigation;
    },
    drawerHeaderNavigationClassNames: function() {
        return _DrawerHeaderNavigation.drawerHeaderNavigationClassNames;
    },
    renderDrawerHeaderNavigation_unstable: function() {
        return _DrawerHeaderNavigation.renderDrawerHeaderNavigation_unstable;
    },
    useDrawerHeaderNavigationStyles_unstable: function() {
        return _DrawerHeaderNavigation.useDrawerHeaderNavigationStyles_unstable;
    },
    useDrawerHeaderNavigation_unstable: function() {
        return _DrawerHeaderNavigation.useDrawerHeaderNavigation_unstable;
    },
    DrawerFooter: function() {
        return _DrawerFooter.DrawerFooter;
    },
    drawerFooterClassNames: function() {
        return _DrawerFooter.drawerFooterClassNames;
    },
    renderDrawerFooter_unstable: function() {
        return _DrawerFooter.renderDrawerFooter_unstable;
    },
    useDrawerFooterStyles_unstable: function() {
        return _DrawerFooter.useDrawerFooterStyles_unstable;
    },
    useDrawerFooter_unstable: function() {
        return _DrawerFooter.useDrawerFooter_unstable;
    }
});
const _Drawer = require("./Drawer");
const _OverlayDrawer = require("./OverlayDrawer");
const _InlineDrawer = require("./InlineDrawer");
const _DrawerBody = require("./DrawerBody");
const _DrawerHeader = require("./DrawerHeader");
const _DrawerHeaderTitle = require("./DrawerHeaderTitle");
const _DrawerHeaderNavigation = require("./DrawerHeaderNavigation");
const _DrawerFooter = require("./DrawerFooter");
